{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# RdRp0 pan-proteome\n",
    "```\n",
    "Lead     : ababaian\n",
    "Issue    : \n",
    "start    : 2020 12 10\n",
    "complete : 2020 12 13\n",
    "revision : 2020 12 15\n",
    "files    : ~/serratus/notebook/201210_rdrp0/\n",
    "s3 files : s3://serratus-public/notebook/201210_rdrp0/\n",
    "```\n",
    "\n",
    "## Introduction\n",
    "\n",
    "We've been considering other ways to 'dive' into the SRA to yield meaningful, interpretable results. An idrea which is recurring is to focus on a gene-family/domain that we would like to characterize exhaustively.\n",
    "\n",
    "The prime candidate is viral RNA-dependent RNA-polymerase or `RdRp`. This is slowly-evolving and central reference gene for the identification and classification of RNA viruses.\n",
    "\n",
    "It is a daunting task to isolate all known RdRp and categorize them into a meaningful system, this is a first approximation of that goal putting together the components to do so.\n",
    "\n",
    "The ideal end-goal will be to create a hierarchly/taxonomically nested set of RdRp protein sequences at various cut-off thresholds.\n",
    "\n",
    "- rdrp100: all unique RdRp sequences\n",
    "- rdrp97:  all rdrp sequences clustered at 97% identity\n",
    "- rdrp90:  Species-Approximate. 90% identity clusters\n",
    "- rdrp75:  Genus-Approximate. 75% identity clusters\n",
    "\n",
    "### Key Literature\n",
    "\n",
    "- [Wolf20: Doubling RNA viruses](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7508674/)\n",
    "- [Wolf18: RdRp evo/origin](https://pubmed.ncbi.nlm.nih.gov/30482837/)\n",
    "- [Venk18: RdRp evo/origin](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5850383/)\n",
    "- [Zhange: Expanding RNA virome (review)](https://pubmed.ncbi.nlm.nih.gov/31100994/)\n",
    "\n",
    "### Objectives\n",
    "- Compile the materials neccesary for a comprehensive RdRp-ome\n",
    "- Create the `rdrp0.fa` reference pan-proteome to run a pilot serratus run and see what results would look like"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sun Dec 13 11:06:21 PST 2020\n",
      "6ac78a036910813c0f5fb2e7ef0b88599e683959\n"
     ]
    }
   ],
   "source": [
    "# Serratus commit version\n",
    "SERRATUS=\"/home/artem/serratus\"\n",
    "cd $SERRATUS\n",
    "\n",
    "# Create local run directory\n",
    "WORK=\"$SERRATUS/notebook/201210_rdrp0\"\n",
    "mkdir -p $WORK; cd $WORK\n",
    "\n",
    "# S3 notebook path\n",
    "S3_WORK='s3://serratus-public/notebook/201210_rdrp0/'\n",
    "\n",
    "# date and version\n",
    "date\n",
    "git rev-parse HEAD # commit version"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GenBank Virome\n",
    "\n",
    "The master corpus for all viral sequences\n",
    "\n",
    "### Nucleotide Sequences\n",
    "- Query: `txid10239[Organism:exp]` # all viruses\n",
    "- Date: `201205`\n",
    "- Results: `3 535 357` sequences\n",
    "- File : `ntViro_gb201005.fa`\n",
    "\n",
    "### CDS Sequences\n",
    "- Query: `txid2552587[Organism:exp]` # all RNA virus CDS\n",
    "- Date: `201205` # error in this version\n",
    "- Results: `2 825 230` sequences\n",
    "- File : `cdsViro_gb201005.fa`\n",
    "\n",
    "### Protein Sequences\n",
    "- Query: `txid10239[Organism:exp]` # all viral protein sequences\n",
    "- Date: `201212` # error in this version\n",
    "- Results: `2 825 230` sequences\n",
    "- File : `aaViro_gb201012.fa`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3535357\n",
      "9102eceda85185cfb124023dbf129621  ntViro_gb201205.fa\n",
      "2825230\n",
      "81946a20fc18b24eb6b49541d41b8dd0  cdsViro_gb201205.fa\n",
      "2825230\n",
      "81946a20fc18b24eb6b49541d41b8dd0  aaViro_gb201212.fa\n",
      "13870\n",
      "671fb5b3f02fd41457be4d6c7a31a417  gbRdRp_201212.fa\n",
      "13870\n",
      "671fb5b3f02fd41457be4d6c7a31a417  gbRdRp_201212.fa\n"
     ]
    }
   ],
   "source": [
    "cd $WORK\n",
    "NT='ntViro_gb201205.fa'\n",
    "grep \">\" $NT | wc -l\n",
    "md5sum $NT\n",
    "md5sum $NT > $NT.md5\n",
    "\n",
    "CDS='cdsViro_gb201205.fa'\n",
    "grep \">\" $CDS | wc -l\n",
    "md5sum $CDS\n",
    "md5sum $CDS > $CDS.md5\n",
    "\n",
    "AA='aaViro_gb201212.fa'\n",
    "grep \">\" $AA | wc -l\n",
    "md5sum $AA\n",
    "md5sum $AA > $AA.md5\n",
    "\n",
    "# GB RdRp sequences (see below WOLF18)\n",
    "GB='gbRdRp_201212.fa'\n",
    "grep \">\" $GB | wc -l\n",
    "md5sum $GB\n",
    "md5sum $GB > $GB.md5\n",
    "\n",
    "# YA RdRp sequences (see below WOLF20)\n",
    "YA='gbRdRp_201212.fa'\n",
    "grep \">\" $YA | wc -l\n",
    "md5sum $YA\n",
    "md5sum $YA > $YA.md5"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## WOLF18 RdRp\n",
    "\n",
    "FTP Access: `ftp://ftp.ncbi.nlm.nih.gov/pub/wolf/_suppl/rnavir18/`\n",
    "\n",
    "Sequence data: `ftp://ftp.ncbi.nlm.nih.gov/pub/wolf/_suppl/rnavir18/RNAvirome.S2.afa`\n",
    "\n",
    "Saved as: `gb_rdrp.afa`\n",
    "\n",
    "Date Accessed: `201212`\n",
    "\n",
    "![Figure 1](/home/artem/serratus/notebook/201210_rdrp0/wolf18/figure1.png)\n",
    "\n",
    "### Level 1 - Supergroup / Branches\n",
    "\n",
    "The RdRp can be broadly classified into 5 branches which will form the lowest level of the hierarchy: `rdrp1`, `rdrp2`, `rdrp3`, `rdrp4`, `rdrp5`.\n",
    "\n",
    ">Branch 1 consists of leviviruses and their eukaryotic relatives, namely, “mitoviruses,” “narnaviruses,” and “ourmiaviruses” (the latter three terms are placed in quotation marks as our analysis contradicts the current ICTV framework, which classifies mitoviruses and narnaviruses as members of one family, Narnaviridae, and ourmiaviruses as members of a free-floating genus, Ourmiavirus).\n",
    "\n",
    "> Branch 2 (“picornavirus supergroup”) consists of a large assemblage of +RNA viruses of eukaryotes, in particular, those of the orders Picornavirales and Nidovirales; the families Caliciviridae, Potyviridae, Astroviridae, and Solemoviridae, a lineage of dsRNA viruses that includes partitiviruses and picobirnaviruses; and several other, smaller groups of +RNA and dsRNA viruses.\n",
    "\n",
    "> Branch 3 consists of a distinct subset of +RNA viruses, including the “alphavirus supergroup” along with the “flavivirus supergroup,” nodaviruses, and tombusviruses; the “statovirus,” “wèivirus,” “yànvirus,” and “zhàovirus” groups; and several additional, smaller groups.\n",
    "\n",
    "> Branch 4 consists of dsRNA viruses, including cystoviruses, reoviruses, and totiviruses and several additional families.\n",
    "\n",
    "> Branch 5 consists of −RNA viruses.\n",
    "\n",
    "Boundary defintions of Branches with relation to RdRp are taken from paper\n",
    "\n",
    "Based on: [Supplementary Data 4](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6282212/bin/mbo006184203sd4.txt)\n",
    "Saved as: `rdrp_representative_branches.tree`\n",
    "\n",
    "### Level 2 - Viral Family\n",
    "\n",
    "Based on: `https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6282212/bin/mbo006184203sd1.xls`\n",
    "Saved as: `wolf18_vlist.xlsx`\n",
    "\n",
    "Spreadsheet includes the fields\n",
    "\n",
    "- RdRp num ID: Ordinal numbering for RdRp\n",
    "- RdRp GenBank Acc: Protein accession ID\n",
    "- NCBI Tax ID: taxid from NCBI\n",
    "- virus name: virus name\n",
    "- taxonomy: taxonomic tree\n",
    "\n",
    "Taxonomy field was parsed to retrieve \"*dae* suffix for \"Family\", relatively appropriate family-name or \"unclassified\" when unavailable. Monkey work.\n",
    "\n",
    "### Level 3 - Sequence/Species\n",
    "\n",
    "Based on: `wolf18_vlist.xlsx`\n",
    "\n",
    "- Virus name field (most shallow taxonomic classification) taken for each record.\n",
    "- GenBank accession taken from each record.\n",
    "\n",
    "### Example RdRp\n",
    "\n",
    "```\n",
    "rdrp5.Hantaviridae.Bowe_virus:AGW23849.1\n",
    "rdrp5.Bunyaviridae.Azagny_virus:AEA42011.1\n",
    "...\n",
    "rdrp2.Coronaviridae.Night_heron_coronavirus_HKU19:YP_005352862.1\n",
    "rdrp2.Coronaviridae.Munia_coronavirus_HKU13_3514:YP_002308505.1\n",
    "rdrp2.Coronaviridae.Wigeon_coronavirus_HKU20:YP_005352870.1\n",
    "rdrp2.Coronaviridae.Feline_infectious_peritonitis_virus:AGZ84535.1\n",
    "rdrp2.Coronaviridae.Lucheng_Rn_rat_coronavirus:YP_009336483.1\n",
    "rdrp2.Coronaviridae.Hipposideros_bat_coronavirus_HKU10:AFU92121.1\n",
    "rdrp2.Coronaviridae.BtMs_AlphaCoV_GS2013:AIA62270.1\n",
    "rdrp2.Coronaviridae.Chaerephon_bat_coronavirus_Kenya_KY41_2006:ADX59465.1\n",
    "rdrp2.Coronaviridae.Porcine_epidemic_diarrhea_virus:AID56804.1\n",
    "rdrp2.Coronaviridae.Bat_coronavirus_CDPHE15_USA_2006:YP_008439200.1\n",
    "rdrp2.Coronaviridae.Anlong_Ms_bat_coronavirus:AID16674.1\n",
    "rdrp2.Coronaviridae.Scotophilus_bat_coronavirus_512:YP_001351683.1\n",
    "rdrp2.Coronaviridae.BtNv_AlphaCoV_SC2013:YP_009201729.1\n",
    "rdrp2.Coronaviridae.Bat_coronavirus_1B:ACA52156.1\n",
    "rdrp2.Coronaviridae.NL63_related_bat_coronavirus:YP_009328933.1\n",
    "rdrp2.Coronaviridae.NL63_related_bat_coronavirus:APD51489.1\n",
    "rdrp2.Coronaviridae.229E_related_bat_coronavirus:ALK43115.1\n",
    "rdrp2.Coronaviridae.Rhinolophus_bat_coronavirus_HKU2:ABQ57223.1\n",
    "rdrp2.Coronaviridae.Wencheng_Sm_shrew_coronavirus:AID16677.1\n",
    "rdrp2.Coronaviridae.Human_coronavirus_HKU1:ABD75543.1\n",
    "rdrp2.Coronaviridae.Betacoronavirus_Erinaceus_VMC_DEU_2012:YP_008719930.1\n",
    "rdrp2.Coronaviridae.Pipistrellus_bat_coronavirus_HKU5:YP_001039961.1\n",
    "rdrp2.Coronaviridae.Rousettus_bat_coronavirus:AOG30811.1\n",
    "rdrp2.Coronaviridae.Rousettus_bat_coronavirus:YP_009273004.1\n",
    "rdrp2.Coronaviridae.Bat_CoV_279_2005:P0C6V9.1\n",
    "rdrp2.Coronaviridae.Bat_Hp_betacoronavirus_Zhejiang2013:YP_009072438.1\n",
    "rdrp2.Coronaviridae.Bottlenose_dolphin_coronavirus_HKU22:AHB63494.1\n",
    "rdrp2.Coronaviridae.Duck_coronavirus:AKF17722.1\n",
    "rdrp2.Coronaviridae.Avian_infectious_bronchitis_virus_partridge_GD_S14_2003:AAT70770.1\n",
    "rdrp2.Coronaviridae.Infectious_bronchitis_virus:ADA83556.1\n",
    "...\n",
    "rdrp1.unclassified.Wenzhou_levi_like_virus_3:APG77299.1\n",
    "rdrp1.Leviviridae.Pseudomonas_phage_PP7:NP_042307.1\n",
    "\n",
    "```\n",
    "\n",
    "saved as: `gb_assign_group.txt`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 201213 REPEAT; accidently duplicated rdrp0.fa output here\n",
    "cd $WORK/wolf18"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">AMN92168.1|Bourbon_virus\n",
      ">YP_009352882.1|Dhori_virus\n",
      ">YP_145794.1|Thogoto_virus\n",
      ">AHB34055.1|Upolu_virus\n",
      ">ABF68025.1|Infectious_salmon_anemia_virus\n",
      ">AQM37684.1|Steelhead_trout_orthomyxovirus_1\n",
      ">APG77864.1|Beihai_orthomyxo_like_virus_2\n",
      ">APG77905.1|Hubei_orthomyxo_like_virus_5\n",
      ">YP_009246481.1|Tilapia_lake_virus\n",
      ">YP_009337891.1|Changping_earthworm_virus_2\n",
      ">rdrp5.unclassified.Wuhan_Insect_virus_3:AJG39263.1\n",
      ">rdrp5.unclassified.Mucorales_RNA_virus_1:AMK47917.1\n",
      ">rdrp5.unclassified.Wenling_crustacean_virus_9:YP_009329879.1\n",
      ">rdrp5.Bunyaviridae.Groundnut_chlorotic_fan_spot_virus:AJT59689.1\n",
      ">rdrp5.Tospoviridae.Soybean_vein_necrosis_virus:ADX01591.1\n",
      ">rdrp5.Tospoviridae.Bean_necrotic_mosaic_virus:YP_006468898.1\n",
      ">rdrp5.Tospoviridae.Polygonum_ringspot_tospovirus:AHZ45965.1\n",
      ">rdrp5.Tospoviridae.Pepper_chlorotic_spot_virus:AQX77525.1\n",
      ">rdrp5.Tospoviridae.Melon_yellow_spot_virus:BAG82842.1\n",
      ">rdrp5.Tospoviridae.Capsicum_chlorosis_virus:AGS78403.1\n"
     ]
    }
   ],
   "source": [
    "grep \">\" gb_rdrp.afa | tail -\n",
    "head gb_assign_group.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "KAPDSAARESLDRASEIMTGKSYNAVHTGDLSKLPNQGESPLRIVDSDLYSERSCCWVIEKEGRVVCKSTTLTRGMTGLLNTTRCSSPSELICKVLTVESLSEKIGDTSVEELLSHGRYFKCALRDQERGKPKSRAIFLSHPFFRLLSSVVETHARSVLSKVSAVYTATASAEQRAMMAAQVVESRKHVLNGDCTKYNEAIDADTLLKVWDAIGMGSIGVMLAYMVRRKCVLIKDTLVECPGGMLMGMFNATATLALQGTTDRFLSFSDDFITSFNSPAELREIEDLLFASCHNLSLKKSYISVASLEINSCTLTRDGDLATGLGCTAGVPFRGPLVTLKQTAAMLSGAVDSGVMPFHSAERLFQIKQQECAYRYNNPTYTTRNEDFLPTCLGGKTVISFQSLLTWDCHPFWYQVHPDGPDTIDQKVLSVLASK\n",
      ">YP_009337891.1\n",
      "WDDQDQSMFLRPKNRTGYGPLIFNTMKRISDMSPTRARELSEVFSVTEKERSISVLASGGTKFVPARGTSVPASTAFWDYQDQMRPIFEHYNIKYTDNSWWHIVICANIFGEYFEILPPTWDRSTLTKLFVEIFSAGLAVKQTEHNRSEGRNIVTMSISLQNFQNFVEEVAKIVNRMTGSHGTDLSSLEKRDLLRKVGLAASIELDTFLASLDKTKWNQLLQISTAMLLLAASYPNDASERRFVLLVGQIWREKCLYFPSKHSYYTGGMKTPKTIDELSRMNDEQLLNDNIRDDLMMVLRHYRKKRVIPQYIKCDLIMLMGMFNHSSTTLHIWPAYANHLDDNQTVSKIIDFCASSDDSMVRAKKILGMSALESYRTISSLWKSMGLNDSEDKSIIHDRLVKVEYNSNVFSMGQLIPNLSRDVAGTKVLYENPEKDLETMKNQLFVYINEGTLSTQDAAIILSDKYLTSLDIHDMLPFQKRHPIFLNNLTSAGLIPQCIPIWCGGTNHIPPELWGTMDDKMYWYHHHKDTGKTNLYLEFLASISTPPDVLASEMPVDFKLGL\n"
     ]
    }
   ],
   "source": [
    "# Rename the header to remove virus name\n",
    "# remove gaps from sequence (unaligned)\n",
    "sed 's/|.*//g' gb_rdrp.afa \\\n",
    "  | sed 's/-//g' - \\\n",
    "  > rdrp_1.tmp\n",
    "\n",
    "# One sequence \"Pseudomonas_phage_phiYY\" has no accession\n",
    "# YP_009618381.1\n",
    "sed -i 's/^>$/>YP_009618381.1/g' rdrp_1.tmp\n",
    "\n",
    "tail -n3 rdrp_1.tmp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mv: missing destination file operand after 'rdrp_3.tmp'\n",
      "Try 'mv --help' for more information.\n"
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "# Iterate through each line / fasta name.\n",
    "# to swap out headers\n",
    "\n",
    "while read -r line; do\n",
    "  # Find headers\n",
    "  if [[ \"$line\" = \">\"* ]]; then\n",
    "    acc=$(echo $line | sed 's/>//g' -)\n",
    "    newheader=$(grep \"$acc\" gb_assign_group.txt)\n",
    "    \n",
    "    if [[ \"$newheader\" = \"\" ]]; then\n",
    "      echo \">NA_$acc\" >> rdrp_2.tmp\n",
    "    else\n",
    "      echo $newheader >> rdrp_2.tmp\n",
    "    fi\n",
    "    \n",
    "  else\n",
    "    # print aa sequence\n",
    "    echo $line >> rdrp_2.tmp \n",
    "  fi\n",
    "done < rdrp_1.tmp\n",
    "\n",
    "\n",
    "# Manually remove first ten sequences (Group II introns outgroup)\n",
    "tail -n +21 rdrp_2.tmp > rdrp_3.tmp\n",
    "\n",
    "mv rdrp_3.tmp > ../gbRdRp_201212.fa\n",
    "rm *.tmp"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## WOLF20 RdRp\n",
    "\n",
    "FTP Access: `ftp://ftp.ncbi.nlm.nih.gov/pub/wolf/_suppl/yangshan/`\n",
    "\n",
    "Sequence data: `gb_rdrp.afa`\n",
    "\n",
    "Saved as: `gb_rdrp.fa`\n",
    "\n",
    "Date Accessed: `201212`\n",
    "\n",
    "![Figure 2](/home/artem/serratus/notebook/201210_rdrp0/wolf20/wolf20_figure2.jpg)\n",
    "\n",
    ">RNA virome analysis performed using complementary DNA derived from approximately 10 l of samples from Yangshan Deep-Water Harbour yielded 4,593 nearly full-length RNA virus RdRPs that formed 2,192 clusters at 75% amino acid identity which represents virus diversity at a level between species and genus. Among the RdRP sequences from GenBank (October 2018), 2,021 comparable clusters were detected. Thus, the 10 l water sample analysed here more than doubles the known diversity of RNA viruses.\n",
    "\n",
    "There are two sets of data treated independently here; the genbank RdRp and the yangshan RdRp. Only yangshen sequences will be considered as the genbank records are not comprehensive. The data is clustered globally at `75%` aa identity.\n",
    "\n",
    "### Level 1/2 - Supergroup / Branches and Clusters\n",
    "\n",
    "From [Supplementary Table 1](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7508674/bin/41564_2020_755_MOESM2_ESM.xlsx) parsed the \"clade id\" field to retrieve branch numbering and relate to the OV.x clusters. A few sequences are unassigned / deep so they will form `rdrp0`.\n",
    "\n",
    "The `OV.x` was renamed to `yaOVx` to remove period. So `OV.1` became `rdrp2.yaOV1`\n",
    "\n",
    "Saved as: `ovx.branches.txt`\n",
    "\n",
    "### Level 3 - ORF Identifier\n",
    "Sequence data is from `rdrp.ya.fa`\n",
    "\n",
    "Header parsing:\n",
    "`>ya20_JAAOEH010000011_1 JAAOEH010000011.1 5194-3782 OV.1 NODE_11_truseq orf.65`\n",
    "will become\n",
    "`>rdrp2.yaOV1.orf65\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mv: missing destination file operand after 'ya_2.tmp'\n",
      "Try 'mv --help' for more information.\n"
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "cd $WORK/wolf20\n",
    "\n",
    "# Iterate through each line / fasta name.\n",
    "# to swap out headers\n",
    "\n",
    "while read -r line; do\n",
    "  # Find headers\n",
    "  if [[ \"$line\" = \">\"* ]]; then\n",
    "    acc=$(echo $line | cut -f 4 -d ' ' - | sed 's/\\.//g' -)\n",
    "    branch=$(grep \"$acc\\.\" ovx.branches.txt)\n",
    "    orf=$(echo $line | sed 's/.* //g' - | sed 's/\\.//g' - )\n",
    "    \n",
    "    if [[ \"$branch\" = \"\" ]]; then\n",
    "      echo \">NA\" >> ya_2.tmp\n",
    "    else\n",
    "      echo \">$branch\"\"$orf\" >> ya_2.tmp\n",
    "    fi\n",
    "    \n",
    "  else\n",
    "    # print aa sequence\n",
    "    echo $line >> ya_2.tmp \n",
    "  fi\n",
    "done < rdrp.ya.fa\n",
    "\n",
    "mv ya_2.tmp ../yaRdRp_201212.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## rdrp0 - pilot panproteome\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8479a3347bbe73224cb2eac0c2138a92  rdrp0.fa\n"
     ]
    }
   ],
   "source": [
    "cd $WORK\n",
    "\n",
    "# Make rdrp0\n",
    "cat gbRdRp_201212.fa yaRdRp_201212.fa > rdrp0.fa\n",
    "samtools faidx rdrp0.fa\n",
    "\n",
    "md5sum rdrp0.fa\n",
    "md5sum rdrp0.fa > rdrp0.fa.md5"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Upload to S3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 17G\n",
      "drwxrwxr-x  4 artem artem 4.0K Dec 13 13:34 \u001b[0m\u001b[01;34m.\u001b[0m\n",
      "drwxr-xr-x 40 artem artem 4.0K Dec 13 13:33 \u001b[01;34m..\u001b[0m\n",
      "-rw-r--r--  1 artem artem 4.0G Dec 12 21:28 aaViro_gb201212.fa\n",
      "-rw-rw-r--  1 artem artem   53 Dec 12 22:35 aaViro_gb201212.fa.md5\n",
      "-rw-r--r--  1 artem artem 4.0G Dec 12 13:33 cdsViro_gb201205.fa\n",
      "-rw-rw-r--  1 artem artem   54 Dec 12 22:35 cdsViro_gb201205.fa.md5\n",
      "-rw-rw-r--  1 artem artem 2.5M Dec 13 13:04 gbRdRp_201212.fa\n",
      "-rw-rw-r--  1 artem artem 8.7G Dec  5 18:20 ntViro_gb201205.fa\n",
      "-rw-rw-r--  1 artem artem   53 Dec 12 22:34 ntViro_gb201205.fa.md5\n",
      "-rw-rw-r--  1 artem artem 4.6M Dec 13 13:34 rdrp0.fa\n",
      "-rw-rw-r--  1 artem artem 534K Dec 13 13:34 rdrp0.fa.fai\n",
      "-rw-rw-r--  1 artem artem   43 Dec 13 13:34 rdrp0.fa.md5\n",
      "drwxrwxr-x  2 artem artem 4.0K Dec 13 13:33 \u001b[01;34mwolf18\u001b[0m\n",
      "drwxrwxr-x  5 artem artem 192K Dec 12 22:20 \u001b[01;34mwolf20\u001b[0m\n",
      "-rw-rw-r--  1 artem artem 2.1M Dec 12 22:14 yaRdRp_201212.fa\n"
     ]
    }
   ],
   "source": [
    "cd $WORK\n",
    "ls -alh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "upload: ./rdrp0.fa.md5 to s3://serratus-public/notebook/201210_rdrp0/rdrp0.fa.md5\n",
      "upload: ./rdrp0.fa.fai to s3://serratus-public/notebook/201210_rdrp0/rdrp0.fa.fai\n",
      "upload: wolf18/rdrp_1.tmp to s3://serratus-public/notebook/201210_rdrp0/wolf18/rdrp_1.tmp\n",
      "upload: wolf18/rdrp_2.tmp to s3://serratus-public/notebook/201210_rdrp0/wolf18/rdrp_2.tmp\n",
      "upload: ./gbRdRp_201212.fa to s3://serratus-public/notebook/201210_rdrp0/gbRdRp_201212.fa\n",
      "upload: ./rdrp0.fa to s3://serratus-public/notebook/201210_rdrp0/rdrp0.fa\n"
     ]
    }
   ],
   "source": [
    "aws s3 sync ./ $S3_WORK # done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create rdrp0 database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load serratus-align container on EC2\n",
    "# From base amazon linux 2\n",
    "sudo yum install -y docker\n",
    "sudo yum install -y git\n",
    "sudo service docker start\n",
    "\n",
    "git clone --branch diamond-dev https://github.com/ababaian/serratus.git; cd serratus\n",
    "\n",
    "# If you want to upload containers to your repository, include this.\n",
    "export DOCKERHUB_USER='serratusbio' # optional\n",
    "sudo docker login # optional\n",
    "\n",
    "# Build all containers and upload them docker hub repo (if available)\n",
    "cd containers\n",
    "./build_containers.sh   # run this in the folder 'serratus/containers'\n",
    "\n",
    "sudo docker run --rm --entrypoint /bin/bash \\\n",
    "  -it serratus-align:latest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# rdrp0 v201213 for pilot run\n",
    "mkdir rdrp0; cd rdrp0\n",
    "GENOME='rdrp0'\n",
    "\n",
    "# Download rdrp0\n",
    "aws s3 cp s3://serratus-public/notebook/201210_rdrp0/rdrp0.fa ./\n",
    "\n",
    "# Make diamond index for protref5\n",
    "diamond makedb --in $GENOME.fa -d $GENOME\n",
    "\n",
    "# Make fasta index for protref5\n",
    "samtools faidx $GENOME.fa\n",
    "mv $GENOME.fa.fai $GENOME.sumzer.tsv\n",
    "\n",
    "md5sum * > $GENOME.md5\n",
    "\n",
    "# use protref5 msa as place-holder\n",
    "aws s3 cp s3://serratus-public/seq/protref5/protref5.msa ./rdrp0.msa\n",
    "\n",
    "# 657d302bd62a9e0b588668101f581e4c  rdrp0.dmnd\n",
    "# 8479a3347bbe73224cb2eac0c2138a92  rdrp0.fa\n",
    "# 6bf2ffa27bb2b08bf0d6056b675fa348  rdrp0.sumzer.tsv\n",
    "# e094fc7db19c07ffcedf8bc42963ab80  rdrp0.msa\n",
    "\n",
    "# Upload to S3\n",
    "aws s3 sync ./ s3://serratus-public/seq/$GENOME/\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Revision 0 - Clean-up \"Virus Name\" field\n",
    "\n",
    "In parsing the \"Virus Name\" from `rdrp0` the \".\" character was in quite a few virus names and these result in icky downstream parsing. This is a manual removal of those periods from all rdrp0 names to make a 'clean' version\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# revision 0 folder\n",
    "mkdir rev0; cd rev0\n",
    "\n",
    "# Wolf18 Genbank Sequences\n",
    "cp ../gbRdRp_201212.fa  ./\n",
    "# Wolf20 Yangshen Sequences\n",
    "cp ../yaRdRp_201212.fa  ./ "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Update Sequence names (to make uniform)\n",
    "# This is from ongoing work from\n",
    "\n",
    "## sp. --> sp\n",
    "grep \"sp\\.\" gbRdRp_201212.fa | less -NS -\n",
    "sed -i 's/sp\\./sp/g' gbRdRp_201212.fa\n",
    "\n",
    "#      1 >rdrp1.Narnaviridae.Rhizophagus_sp._HR1_mitovirus_like_ssRNA:BAN85985.1\n",
    "#      2 >rdrp1.Narnaviridae.Rhizophagus_sp._RF1_mitovirus:BAJ23143.2\n",
    "#      3 >rdrp2.Picobirnaviridae.Picobirnavirus_sp.:AQS16638.1\n",
    "#      4 >rdrp2.Picobirnaviridae.Picobirnavirus_sp.:AOW41971.1\n",
    "#      5 >rdrp2.Picobirnaviridae.Picobirnavirus_sp.:AOW41973.1\n",
    "#      6 >rdrp2.Iflaviridae.Iflavirus_sp.:APB88805.1\n",
    "#      7 >rdrp2.unclassified.Chaetoceros_sp._RNA_virus_2:BAK40203.1\n",
    "#      8 >rdrp2.unclassified.Posavirus_sp.:APQ44560.1\n",
    "#      9 >rdrp2.unclassified.Posavirus_sp.:APQ44553.1\n",
    "#     10 >rdrp2.unclassified.Posavirus_sp.:APQ44556.1\n",
    "#     11 >rdrp2.unclassified.Posavirus_sp.:APQ44559.1\n",
    "#     12 >rdrp2.unclassified.Posavirus_sp.:APQ44517.1\n",
    "#     13 >rdrp2.unclassified.Basavirus_sp.:APQ44489.1\n",
    "#     14 >rdrp2.unclassified.Basavirus_sp.:APQ44495.1\n",
    "#     15 >rdrp2.unclassified.Posavirus_sp.:APQ44558.1\n",
    "#     16 >rdrp2.unclassified.Basavirus_sp.:APQ44499.1\n",
    "#     17 >rdrp2.unclassified.Basavirus_sp.:APQ44492.1\n",
    "#     18 >rdrp2.unclassified.Basavirus_sp.:APQ44496.1\n",
    "#     19 >rdrp2.unclassified.Basavirus_sp.:APQ44502.1\n",
    "#     20 >rdrp2.unclassified.Posavirus_sp.:APQ44531.1\n",
    "#     21 >rdrp2.unclassified.Posavirus_sp.:YP_009333148.1\n",
    "#     22 >rdrp2.unclassified.Posavirus_sp.:APQ44537.1\n",
    "#     23 >rdrp2.unclassified.Rasavirus_sp.:APQ44507.1\n",
    "#     24 >rdrp2.unclassified.Husavirus_sp.:APQ44514.1\n",
    "#     25 >rdrp2.unclassified.Rasavirus_sp.:APQ44506.1\n",
    "#     26 >rdrp2.unclassified.Rasavirus_sp.:YP_009333305.1\n",
    "#     27 >rdrp2.unclassified.Posavirus_sp.:APQ44547.1\n",
    "#     28 >rdrp2.unclassified.Basavirus_sp.:APQ44500.1\n",
    "#     29 >rdrp2.Picornaviridae.Sicinivirus_sp.:APR73491.1\n",
    "#     30 >rdrp2.Picornaviridae.Picornaviridae_sp._rodent_Ee_PicoV_NX2015:APA29022.1\n",
    "#     31 >rdrp2.Picornaviridae.Picornaviridae_sp._rodent_Mc_PicoV_Tibet2015:APA29023.1\n",
    "#     32 >rdrp2.Picornaviridae.Enterovirus_sp.:AHY21610.1\n",
    "#     33 >rdrp2.Picornaviridae.Picornaviridae_sp._rodent_CK_PicoV_Tibet2014:APA29019.1\n",
    "#     34 >rdrp2.Picornaviridae.Picornaviridae_sp._rodent_Rn_PicoV_SX2015_1:APA29018.1\n",
    "#     35 >rdrp2.Picornaviridae.Picornaviridae_sp._rodent_Ds_PicoV_IM2014:APA29017.1\n",
    "#     36 >rdrp3.Picobirnaviridae.Picobirnavirus_sp.:AOW41972.1\n",
    "#     37 >rdrp4.unclassified.unclassified_Rhizophagus_sp._RF1_medium_virus:BAJ23141.1\n",
    "#     38 >rdrp4.Chrysoviridae.Fusarium_oxysporum_f._sp._dianthi_mycovirus_1:YP_009158913.1\n",
    "#     39 >rdrp5.Bunyaviridae.Bunyavirus_sp.:AOY18806.1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Manual interventions\n",
    "\n",
    "grep \">\" gbRdRp_201212.fa | cut -f3- -d'.' | cut -f1 -d':' | grep \"\\.\" - | less -NS\n",
    "\n",
    "#      1 Chaetoceros_socialis_f._radians_RNA_virus_01\n",
    "sed -i 's/f\\./f/g' gbRdRp_201212.fa\n",
    "#      2 Norovirus_dog_GVI.1_HKU_Ca026F_2007_HKG\n",
    "sed -i 's/GVI\\.1/GVI_1/g' gbRdRp_201212.fa\n",
    "#      3 Norovirus_cat_GIV.2_CU081210E_USA_2010\n",
    "sed -i 's/GIV\\.2/GIV_2/g' gbRdRp_201212.fa\n",
    "#      4 Norovirus_Hu_GII.12_CGMH42_2010_TW\n",
    "#      5 Norovirus_Hu_GII.12_CGMH40_2010_TW\n",
    "#      6 Norovirus_GII.17\n",
    "sed -i 's/GII\\.12/GII_12/g' gbRdRp_201212.fa\n",
    "sed -i 's/GII\\.17/GII_17/g' gbRdRp_201212.fa\n",
    "#      7 Sapovirus_Hu_GI.2_BR_DF01_BRA_2009\n",
    "sed -i 's/GI\\.2/GI_2/g' gbRdRp_201212.fa\n",
    "#      8 Sapovirus_GII.8\n",
    "sed -i 's/GII\\.8/GII_8/g' gbRdRp_201212.fa\n",
    "#      9 St._Louis_encephalitis_virus\n",
    "#     10 St._Louis_encephalitis_virus\n",
    "sed -i 's/St\\._L/St_L/g' gbRdRp_201212.fa\n",
    "#     11 Fusarium_oxysporum_f._sp_dianthi_mycovirus_1\n",
    "##\n",
    "\n",
    "# Accesion field does not contain a version number\n",
    "# Do not chnage accessions at this point\n",
    "grep \">\" gbRdRp_201212.fa |  grep -v \"\\.[0-9]$\" -\n",
    "#>rdrp1.Leviviridae.Escherichia_virus_Qbeta:4R71\n",
    "#>rdrp3.Flaviviridae.Douroucouli_hepatitis_GB_virus_A:T08841\n",
    "#>rdrp3.Flaviviridae.Marmoset_hepatitis_GB_virus_A:T08839\n",
    "#>rdrp3.Alphaflexiviridae.Potato_aucuba_mosaic_virus:2012194A\n",
    "#>rdrp3.Virgaviridae.Barley_stripe_mosaic_virus:2211403A\n",
    "#>rdrp4.Reoviridae.Simian_rotavirus:2R7Q\n",
    "#>rdrp4.unclassified.White_button_mushroom_virus_1:T00494\n",
    "#>rdrp5.Peribunyaviridae.La_Crosse_virus:5AMR_A\n",
    "#>rdrp5.Orthomyxoviridae.Influenza_B_virus_(B_Memphis_13_2003):4WRT_B\n",
    "#>rdrp5.Orthomyxoviridae.Influenza_A_virus_(A_little_yellow_shouldered_bat_Guatemala_060_2010(H17N10)):4WSB_B\n",
    "\n",
    "\n",
    "md5sum *\n",
    "# 7d2c1858d5d842ada3e8103c43f29e27  gbRdRp_201212.fa\n",
    "# 7e6781ed5902ea6034393c0dae521c62  yaRdRp_201212.fa\n",
    "\n",
    "# Overwrite previous gb versions\n",
    "mv gbRdRp_201212.fa ../\n",
    "mv yaRdRp_201212.fa ../"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Revision 1 - Cluster known + unclassified sequences\n",
    "\n",
    "Some lessons learned so far... there are ALOT of new viruses to be discovered! From the ~9.5K datasets run with the `rdrp0` pilot, there are ~4,700 high score (>50) and high divergence (55-85% aa id) hits at the family level. Random sampling from vertebrate/virome SRA queries means overlap is likely to be minimal here but a conservative estimate would be 1,000 distinct RdRp, about 25% of the known biodiversity of sequences.\n",
    "\n",
    "### Objectives\n",
    "\n",
    "- There is a large amount of \"unclassified\" sequences in the data within branches, these should be grouped relative to one another such that if two unclassified sequences exist in one branch, they do not \"collide\" in the read summaries.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tue Dec 15 20:57:00 PST 2020\n",
      "6ac78a036910813c0f5fb2e7ef0b88599e683959\n"
     ]
    }
   ],
   "source": [
    "# Serratus commit version\n",
    "SERRATUS=\"/home/artem/serratus\"\n",
    "cd $SERRATUS\n",
    "\n",
    "# Create local run directory\n",
    "WORK=\"$SERRATUS/notebook/201210_rdrp0\"\n",
    "mkdir -p $WORK; cd $WORK\n",
    "\n",
    "# S3 notebook path\n",
    "S3_WORK='s3://serratus-public/notebook/201210_rdrp0/'\n",
    "\n",
    "# date and version\n",
    "date\n",
    "git rev-parse HEAD # commit version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Work on EC2 Instance\n",
    "\n",
    "# Local usearch install\n",
    "#The clustered database was made with usearch:\n",
    "wget https://drive5.com/downloads/usearch11.0.667_i86linux32.gz\n",
    "gzip -dc usearch11.0.667_i86linux32.gz > usearch\n",
    "chmod 755 usearch; mv usearch /usr/bin/usearch\n",
    "\n",
    "# Install seqkit\n",
    "wget https://github.com/shenwei356/seqkit/releases/download/v0.12.0/seqkit_linux_amd64.tar.gz\n",
    "  tar -xvf seqkit*\n",
    "  sudo mv seqkit /usr/local/bin/\n",
    "  rm seqkit_linux*\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize workspace\n",
    "\n",
    "# Download rdrp working directory\n",
    "mkdir rdrp0; cd rdrp0\n",
    "aws s3 sync s3://serratus-public/notebook/201210_rdrp0/ ./\n",
    "\n",
    "# revision 1 folder\n",
    "mkdir rev1; cd rev1\n",
    "\n",
    "# Wolf18 Genbank Sequences\n",
    "cp ../gbRdRp_201212.fa  ./\n",
    "# Wolf20 Yangshen Sequences\n",
    "cp ../yaRdRp_201212.fa  ./ \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# in gbRdRp; seperate out taxonomic and unclassified sequences\n",
    "grep \">\" gbRdRp_201212.fa | wc -l\n",
    "# 4617\n",
    "\n",
    "# -v inverse match\n",
    "seqkit grep -v -r -p 'unclassified' gbRdRp_201212.fa > gb_tax.fa\n",
    "grep \">\" gb_tax.fa | wc -l\n",
    "# 2863\n",
    "\n",
    "seqkit grep -r -p 'unclassified' gbRdRp_201212.fa > gb_unc.fa\n",
    "grep \">\" gb_unc.fa | wc -l\n",
    "# 1754\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sort and Uclust unclassified sequences\n",
    "\n",
    "usearch -sortbylength gb_tax.fa \\\n",
    "   -fastaout gb_tax.sort.fa\n",
    "\n",
    "usearch -sortbylength gb_unc.fa \\\n",
    "   -fastaout gb_unc.sort.fa\n",
    "\n",
    "cat gb_tax.sort.fa gb_unc.sort.fa > gb_cat.sort.fa\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Cluster sequences at 75% identity\n",
    "\n",
    "# Prune TAX\n",
    "usearch -cluster_smallmem gb_tax.sort.fa \\\n",
    "   -id 0.75 \\\n",
    "   -maxaccepts 4 \\\n",
    "   -maxrejects 64 \\\n",
    "   -maxhits 1 \\\n",
    "   -uc gb_tax.id75.uc \\\n",
    "   -centroids gb_tax.id75.fa\n",
    "\n",
    "#      Seqs  2863\n",
    "#  Clusters  1799\n",
    "#  Max size  29\n",
    "#  Avg size  1.6\n",
    "#  Min size  1\n",
    "# Singletons  1353, 47.3% of seqs, 75.2% of clusters\n",
    "\n",
    "\n",
    "# Prune UNC\n",
    "usearch -cluster_smallmem gb_unc.sort.fa \\\n",
    "   -id 0.75 \\\n",
    "   -maxaccepts 4 \\\n",
    "   -maxrejects 64 \\\n",
    "   -maxhits 1 \\\n",
    "   -uc gb_unc.id75.uc \\\n",
    "   -centroids gb_unc.id75.fa\n",
    "\n",
    "#  Seqs      1754\n",
    "#  Clusters  1660\n",
    "#  Max size  6\n",
    "#  Avg size  1.1\n",
    "#  Min size  1\n",
    "# Singletons  1581, 90.1% of seqs, 95.2% of clusters\n",
    "\n",
    "\n",
    "# Prune CAT\n",
    "usearch -cluster_smallmem gb_cat.sort.fa \\\n",
    "   -sortedby other \\\n",
    "   -id 0.75 \\\n",
    "   -maxaccepts 4 \\\n",
    "   -maxrejects 64 \\\n",
    "   -maxhits 1 \\\n",
    "   -uc gb_cat.id75.uc \\\n",
    "   -centroids gb_cat.id75.fa\n",
    "   \n",
    "#      Seqs  4617\n",
    "#  Clusters  3433\n",
    "#  Max size  29\n",
    "#  Avg size  1.3\n",
    "#  Min size  1\n",
    "# Singletons  2892, 62.6% of seqs, 84.2% of clusters\n",
    "\n",
    "\n",
    "## At 75%, no \"unclassifed\" sequence groups with a taxonomic identifier\n",
    "\n",
    "mkdir id75\n",
    "mv *id75.* id75/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Repeat process at 55%; base of diamond detection\n",
    "\n",
    "# Prune TAX\n",
    "usearch -cluster_smallmem gb_tax.sort.fa \\\n",
    "   -id 0.55 \\\n",
    "   -maxaccepts 4 \\\n",
    "   -maxrejects 64 \\\n",
    "   -maxhits 1 \\\n",
    "   -uc gb_tax.id55.uc \\\n",
    "   -centroids gb_tax.id55.fa\n",
    "\n",
    "#      Seqs  2863\n",
    "#  Clusters  828\n",
    "#  Max size  115\n",
    "#  Avg size  3.5\n",
    "#  Min size  1\n",
    "#  Singletons  466, 16.3% of seqs, 56.3% of clusters\n",
    "\n",
    "\n",
    "# Prune UNC\n",
    "usearch -cluster_smallmem gb_unc.sort.fa \\\n",
    "   -id 0.55 \\\n",
    "   -maxaccepts 4 \\\n",
    "   -maxrejects 64 \\\n",
    "   -maxhits 1 \\\n",
    "   -uc gb_unc.id55.uc \\\n",
    "   -centroids gb_unc.id55.fa\n",
    "\n",
    "#      Seqs  1754\n",
    "#  Clusters  1395\n",
    "#  Max size  10\n",
    "#  Avg size  1.3\n",
    "#  Min size  1\n",
    "#  Singletons  1171, 66.8% of seqs, 83.9% of clusters\n",
    "\n",
    "\n",
    "# Prune CAT\n",
    "usearch -cluster_smallmem gb_cat.sort.fa \\\n",
    "   -sortedby other \\\n",
    "   -id 0.55 \\\n",
    "   -maxaccepts 4 \\\n",
    "   -maxrejects 64 \\\n",
    "   -maxhits 1 \\\n",
    "   -uc gb_cat.id55.uc \\\n",
    "   -centroids gb_cat.id55.fa\n",
    "\n",
    "#      Seqs  4617\n",
    "#  Clusters  2150\n",
    "#  Max size  115\n",
    "#  Avg size  2.1\n",
    "#  Min size  1\n",
    "#  Singletons  1552, 33.6% of seqs, 72.2% of clusters\n",
    "\n",
    "## At 55%, no \"unclassifed\" sequence groups with a taxonomic identifier\n",
    "\n",
    "mkdir id55\n",
    "mv *.id55.* id55/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Repeat process at 45%; base of diamond detection\n",
    "\n",
    "# Prune TAX\n",
    "usearch -cluster_smallmem gb_tax.sort.fa \\\n",
    "   -id 0.45 \\\n",
    "   -maxaccepts 4 \\\n",
    "   -maxrejects 64 \\\n",
    "   -maxhits 1 \\\n",
    "   -uc gb_tax.id45.uc \\\n",
    "   -centroids gb_tax.id45.fa\n",
    "\n",
    "#      Seqs  2863\n",
    "#  Clusters  598\n",
    "#  Max size  155\n",
    "#  Avg size  4.8\n",
    "#  Min size  1\n",
    "#  Singletons  302, 10.5% of seqs, 50.5% of clusters\n",
    "\n",
    "\n",
    "# Prune UNC\n",
    "usearch -cluster_smallmem gb_unc.sort.fa \\\n",
    "   -id 0.45 \\\n",
    "   -maxaccepts 4 \\\n",
    "   -maxrejects 64 \\\n",
    "   -maxhits 1 \\\n",
    "   -uc gb_unc.id45.uc \\\n",
    "   -centroids gb_unc.id45.fa\n",
    "\n",
    "#      Seqs  1754\n",
    "#  Clusters  1158\n",
    "#  Max size  17\n",
    "#  Avg size  1.5\n",
    "#  Min size  1\n",
    "#  Singletons  875, 49.9% of seqs, 75.6% of clusters\n",
    "\n",
    "\n",
    "# Prune CAT\n",
    "usearch -cluster_smallmem gb_cat.sort.fa \\\n",
    "   -sortedby other \\\n",
    "   -id 0.45 \\\n",
    "   -maxaccepts 4 \\\n",
    "   -maxrejects 64 \\\n",
    "   -maxhits 1 \\\n",
    "   -uc gb_cat.id45.uc \\\n",
    "   -centroids gb_cat.id45.fa\n",
    "\n",
    "#      Seqs  4617\n",
    "#  Clusters  1653\n",
    "#  Max size  155\n",
    "#  Avg size  2.8\n",
    "#  Min size  1\n",
    "#  Singletons  1068, 23.1% of seqs, 64.6% of clusters\n",
    "\n",
    "\n",
    "## At 45%, no \"unclassifed\" sequence groups with a taxonomic identifier.\n",
    "\n",
    "mkdir id45\n",
    "mv *.id45.* id45/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Repeat process at 35%; base of diamond detection\n",
    "\n",
    "# Prune TAX\n",
    "usearch -cluster_smallmem gb_tax.sort.fa \\\n",
    "   -id 0.35 \\\n",
    "   -maxaccepts 4 \\\n",
    "   -maxrejects 64 \\\n",
    "   -maxhits 1 \\\n",
    "   -uc gb_tax.id35.uc \\\n",
    "   -centroids gb_tax.id35.fa\n",
    "\n",
    "#      Seqs  2863\n",
    "#  Clusters  548\n",
    "#  Max size  156\n",
    "#  Avg size  5.2\n",
    "#  Min size  1\n",
    "#  Singletons  259, 9.0% of seqs, 47.3% of clusters\n",
    "\n",
    "\n",
    "# Prune UNC\n",
    "usearch -cluster_smallmem gb_unc.sort.fa \\\n",
    "   -id 0.35 \\\n",
    "   -maxaccepts 4 \\\n",
    "   -maxrejects 64 \\\n",
    "   -maxhits 1 \\\n",
    "   -uc gb_unc.id35.uc \\\n",
    "   -centroids gb_unc.id35.fa\n",
    "\n",
    "#      Seqs  1754\n",
    "#  Clusters  1097\n",
    "#  Max size  17\n",
    "#  Avg size  1.6\n",
    "#  Min size  1\n",
    "#  Singletons  801, 45.7% of seqs, 73.0% of clusters\n",
    "\n",
    "\n",
    "# Prune CAT\n",
    "usearch -cluster_smallmem gb_cat.sort.fa \\\n",
    "   -sortedby other \\\n",
    "   -id 0.35 \\\n",
    "   -maxaccepts 4 \\\n",
    "   -maxrejects 64 \\\n",
    "   -maxhits 1 \\\n",
    "   -uc gb_cat.id35.uc \\\n",
    "   -centroids gb_cat.id35.fa\n",
    "\n",
    "#      Seqs  4617\n",
    "#  Clusters  1528\n",
    "#  Max size  156\n",
    "#  Avg size  3.0\n",
    "#  Min size  1\n",
    "#  Singletons  941, 20.4% of seqs, 61.6% of clusters\n",
    "\n",
    "\n",
    "## At 35%, no \"unclassifed\" sequence groups with a taxonomic identifier.\n",
    "\n",
    "mkdir id35\n",
    "mv *.id35.* id35/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mkdir -p all\n",
    "mv *.fa all/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Conclusions\n",
    "\n",
    "The clustering here is actually not as extensive as I originally had thought it would be. It looks like "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Assigning Sequence Clusters\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Identity to use for clustering\n",
    "ID='45'\n",
    "\n",
    "# Extract centroid sequence names\n",
    "grep \">\" id$ID/gb_cat.id$ID.fa \\\n",
    "  | sed 's/>//g' \\\n",
    "  > centroid$ID.name.tmp\n",
    "\n",
    "\n",
    "# Convert uc format to TSV\n",
    "# CENTROID hit1 hit2 hitN ...\n",
    "rm clust.tmp; touch clust.tmp\n",
    "\n",
    "# Read through centroid names\n",
    "while read -r line; do\n",
    "  # parse to just accession number\n",
    "  acc=$(echo $line | sed 's/.*://g')\n",
    "  \n",
    "  # Search accession in uclust file\n",
    "  # extract seqname and make it one line\n",
    "  grep \"$acc\" id$ID/gb_cat.id$ID.uc \\\n",
    "    | cut -f 9 \\\n",
    "    | tr '\\n' '\\t' \\\n",
    "    >> clust.tmp\n",
    "    \n",
    "  echo -e \"\\n\" >> clust.tmp\n",
    "\n",
    "done < centroid$ID.name.tmp\n",
    "\n",
    "# remove empty rows\n",
    "cat clust.tmp | sed '/^$/d' > cluster$ID.tsv\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Assign \"family_id\" to sequences (unclassified)\n",
    "\n",
    "# Field 1: Centroid Name\n",
    "# Field 2+: Sequence name\n",
    "cut -f 1 cluster$ID.tsv \\\n",
    "  | sed 's/\\./\\t/g' - \\\n",
    "  | sed 's/:/\\t/g' - \\\n",
    "  > centroid$ID.info\n",
    "\n",
    "# Centroid Information tsv\n",
    "# 1: Branch\n",
    "# 2: Family\n",
    "# 3: Rep. Viral name\n",
    "# 4: Rep. Viral accession\n",
    "# 5: Rep. Viral accession version\n",
    "# 6: Re-annotated Viral Family Name\n",
    "\n",
    "# Iterate through centroid file\n",
    "# assign each \"unclassified family\" a unique ordinal number\n",
    "# 1055 --> use 4 leading zeros\n",
    "\n",
    "# ID\n",
    "N=1\n",
    "rm family2.tmp\n",
    "\n",
    "while read -r line; do\n",
    "  # Read Family name\n",
    "  #echo $line\n",
    "  \n",
    "  family=$(echo $line | cut -f 2 -d' ' -)  \n",
    "\n",
    "  if [[ \"$family\" = \"unclassified\" ]]; then\n",
    "\n",
    "    uncN=$(printf \"%04d\" $N)\n",
    "    echo \"unc$uncN\" >> family2.tmp\n",
    "    \n",
    "    # increment up\n",
    "    N=$((N+1))\n",
    "    \n",
    "  else\n",
    "    echo $family >> family2.tmp\n",
    "  \n",
    "  fi\n",
    "\n",
    "done < centroid$ID.info\n",
    "\n",
    "# Add 6th column of new family names\n",
    "cp centroid$ID.info centroid.tmp\n",
    "paste family2.tmp centroid.tmp  > centroid$ID.info\n",
    "rm *.tmp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Assign Centroid Family-Name to Each Sequence\n",
    "cat all/gbRdRp_201212.fa all/yaRdRp_201212.fa > rdrp0_r1.fa\n",
    "\n",
    "#\n",
    "cut -f 1  centroid$ID.info > fam.name.tmp\n",
    "cut -f 2- cluster$ID.tsv   > cluster.members.tmp\n",
    "paste fam.name.tmp cluster.members.tmp > assign.family.tmp\n",
    "# col 1 == new name\n",
    "# col N == old sequence name\n",
    "\n",
    "while read -r line; do\n",
    "  newfam=$(echo $line | cut -f1 -d' ' -)\n",
    "  echo $newfame\n",
    "  \n",
    "  echo $line \\\n",
    "    | cut -f 2- -d' ' - \\\n",
    "    > members.tmp\n",
    "    \n",
    "  cat members.tmp \\\n",
    "    | tr \" \" \"\\n\" \\\n",
    "    > members2.tmp\n",
    "    \n",
    "    while read -r line2; do\n",
    "      # rdrp5.Sunviridae.Sunshine_Coast_virus:YP_009094051.1\n",
    "      branch=$(echo $line2 | cut -f1 -d'.')\n",
    "      oldfam=$(echo $line2 | cut -f2 -d'.')\n",
    "      virname=$(echo $line2 | cut -f3 -d'.' | cut -f1 -d':')\n",
    "      acc=$(echo $line2 | cut -f2 -d':')\n",
    "      \n",
    "      echo $branch $oldfam $virname $acc\n",
    "      echo $branch $newfam $virname $acc\n",
    "      \n",
    "      # Inline rename\n",
    "      matchline=$( echo $(grep -n $acc rdrp0_r1.fa  | cut -f1 -d':' -)s)\n",
    "      \n",
    "      sed -i \"$matchline/.*/>$branch.$newfam.$virname:$acc/\" rdrp0_r1.fa\n",
    "      \n",
    "    done < members2.tmp\n",
    "    \n",
    "    echo ''\n",
    "\n",
    "done < assign.family.tmp\n",
    "\n",
    "grep \">\" rdrp0_r1.fa | sed 's/>//g' > rdrp0_r1.fai\n",
    "\n",
    "rm *.tmp\n",
    "rm gbRdRp.fa yaRdRp.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Revision 2 - RdRp from GenBank"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mkdir rev2; cd rev2\n",
    "cp ../rdrp0.fa ./"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Install diamond\n",
    "# DIAMONDVERSION='0.9.35'\n",
    "# wget --quiet https://github.com/bbuchfink/diamond/releases/download/v\"$DIAMONDVERSION\"/diamond-linux64.tar.gz\n",
    "# tar -xvf diamond-linux64.tar.gz\n",
    "# rm diamond-linux64.tar.gz\n",
    "# sudo mv    diamond /usr/local/bin/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Install diamond\n",
    "DIAMONDVERSION='2.0.6-dev'\n",
    "cd ~/\n",
    "\n",
    "# Libraries for diamond\n",
    "yum -y install gcc gcc-c++ glibc-devel \\\n",
    "  cmake  patch automake zlib-devel\n",
    "\n",
    "# grab latest with fix from Benjamin\n",
    "git clone https://github.com/bbuchfink/diamond.git\n",
    "cd diamond\n",
    "\n",
    "mkdir bin; cd bin\n",
    "cmake ..\n",
    "make -j4\n",
    "sudo make install\n",
    "\n",
    "# build\n",
    "# cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_BUILD_MARCH=nehalem ..\n",
    "# make && make install"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "GENOME='rdrp0'\n",
    "cp ../rev1/rdrp0_r1.fa ./rdrp0.fa\n",
    "\n",
    "diamond makedb --in $GENOME.fa -d $GENOME"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Link to all genbank CDS\n",
    "ln -s ../ntViro_gb201205.fa ./\n",
    "\n",
    "# Input file\n",
    "IN='ntViro_gb201205.fa'\n",
    "#IN='tmp.fa'\n",
    "\n",
    "GENOME='rdrp0'\n",
    "\n",
    "# Output name\n",
    "OUTNAME='gbViro_rdrp'\n",
    "\n",
    "# Diamond blastx alignment\n",
    "time cat $IN |\\\n",
    "diamond blastx \\\n",
    "  -d \"$GENOME\".dmnd \\\n",
    "  --unal 0 \\\n",
    "  -k 1 \\\n",
    "  -p 4 \\\n",
    "  -b 1 \\\n",
    "  -f 6 qseqid qstart qend qlen qstrand \\\n",
    "       sseqid sstart send slen \\\n",
    "       pident evalue cigar \\\n",
    "       qseq qseq_translated \\\n",
    "  > \"$OUTNAME\".pro\n",
    "  \n",
    "  \n",
    "# real    227m38.435s\n",
    "# user    899m1.421s\n",
    "# sys     0m17.285s\n",
    "\n",
    "\n",
    "# qseq\n",
    "# qseq_translated \n",
    "# full_qseq_mate\n",
    "\n",
    "# tmp.fa timing tests:\n",
    "# tail -n +20000000 ntViro_gb201205.fa | head -n 500000 > tmp.fa\n",
    "\n",
    "# default\n",
    "# real    5m22.409s\n",
    "# user    21m6.984s\n",
    "# sys     0m0.410s\n",
    "\n",
    "# --sensitive\n",
    "# real    11m19.287s\n",
    "# user    44m18.393s\n",
    "# sys     0m0.595s\n",
    "\n",
    "# output size is equal\n",
    "\n",
    "# NOTE: sseq returns protein sequence from the database, not the query\n",
    "# Changed field 15 sseq to qseq\n",
    "# use qseq instead"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# CDS of results\n",
    "cut -f1,14 $OUTNAME.pro |  sed 's/^/>/g' | sed 's/\\t/\\n/g' > $OUTNAME.cds_hit.fa\n",
    "\n",
    "# AA of results\n",
    "cut -f1,15 $OUTNAME.pro |  sed 's/^/>/g' | sed 's/\\t/\\n/g' > $OUTNAME.aa_hit.fa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# UCLUST\n",
    "INPUT=\"$OUTNAME.aa_hit.fa\"\n",
    "OUTPUT=\"$OUTNAME.aa_hit.id75.fa\"\n",
    "\n",
    "usearch -sortbylength $INPUT \\\n",
    "   -fastaout tmp.sort.fa\n",
    "\n",
    "\n",
    "# Prune UNC\n",
    "usearch -cluster_smallmem tmp.sort.fa \\\n",
    "   -id 0.75 \\\n",
    "   -maxaccepts 4 \\\n",
    "   -maxrejects 64 \\\n",
    "   -maxhits 1 \\\n",
    "   -uc $OUTPUT.uc \\\n",
    "   -centroids $OUTPUT\n",
    "\n",
    "#      Seqs  318071 (318.1k)\n",
    "#  Clusters  4081\n",
    "#  Max size  75551 (75.6k)\n",
    "#  Avg size  77.9\n",
    "#  Min size  1\n",
    "\n",
    "# UCLUST\n",
    "OUTPUT=\"$OUTNAME.aa_hit.id85.fa\"\n",
    "\n",
    "# Prune UNC\n",
    "usearch -cluster_smallmem tmp.sort.fa \\\n",
    "   -id 0.85 \\\n",
    "   -maxaccepts 4 \\\n",
    "   -maxrejects 64 \\\n",
    "   -maxhits 1 \\\n",
    "   -uc $OUTPUT.uc \\\n",
    "   -centroids $OUTPUT\n",
    "\n",
    "#      Seqs  318071 (318.1k)\n",
    "#  Clusters  4748\n",
    "#  Max size  75528 (75.5k)\n",
    "#  Avg size  67.0\n",
    "#  Min size  1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# From base amazon linux 2\n",
    "sudo yum install -y docker\n",
    "sudo yum install -y git\n",
    "sudo service docker start\n",
    "sudo docker run --rm --entrypoint /bin/bash -it serratus-align:latest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# HMMER\n",
    "wget http://eddylab.org/software/hmmer/hmmer-3.3.1.tar.gz ./"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
