{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# rdrp Benchmark / GITB\n",
    "```\n",
    "Lead     : ababaian /rce\n",
    "Issue    : \n",
    "start    : 2020 12 10\n",
    "complete : 2020 12 13\n",
    "revision : 2020 12 15\n",
    "files    : ~/serratus/notebook/201210_rdrp0/\n",
    "s3 files : s3://serratus-public/notebook/201210_rdrp0/\n",
    "```\n",
    "\n",
    "## Introduction\n",
    "\n",
    "> `rce:`\n",
    "> For validation of RdRp identification and trimming, we should do hold-out cross-validation using high-quality sequences, i.e. RefSeqs. To assess how accuracy varies with identity, we will use cross-validation by identity (CVI, https://www.drive5.com/usearch/manual/cvi.html) which constructs test/training sets where the maximum identity (MaxID) of the top test-training hit is fixed to a known value or range of values. Here, I suggest five test/training pairs with MaxID = 90, 75, 50, 30, 15% to get a representative range.\n",
    ">\n",
    "> The gold standard reference for the gene identification and trimming benchmark (GITB-gold) will be full-length RdRp-containing ORFs from RefSeqs plus coordinates of the trimmed RdRps. These coordinates will be obtained by aligning the ORFs to Wolf18. These alignments should be good (diamond may be good enough) because the top-hit identities should be at least 90%.\n",
    ">\n",
    "> The decoy reference (GITB-decoy) will be all RefSeq ORFs which do NOT contain RdRp. Ideally, these will all be discarded by the protocol. In practice, there will be FPs due to local homology between RdRp and other genes; this surely arises often in viruses due to recombination events. Using the decoy reference will enable us to design better filters and measure the FP rate (ROC curve) of the protocol.\n",
    "\n",
    "Most of the code for this was developed already in `201210_RdRp_panproteome_v1.ipyn` notebook. This is a self-contained migration of that code / clean-up."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wed Dec 30 15:25:39 PST 2020\n"
     ]
    }
   ],
   "source": [
    "# date and version\n",
    "date"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Upload gitb folder (from EC2)\n",
    "aws s3 sync ./ s3://serratus-public/notebook/201230_gitb/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Materials\n",
    "\n",
    "- `wolf18.fa` : Trimmed Set of RdRp True Positives\n",
    "```\n",
    "# Header fields:\n",
    "# rdrp_branch.viral_family.viral_name:accession\n",
    "```\n",
    "\n",
    "- `vrs240.fa` : Viral RefSeq Sequences (aa)\n",
    "\n",
    "\n",
    "\n",
    "- `vrs240_w18.fa` : ALL RefSeq Sequences with a wolf18 match (includes low confidence)\n",
    "\n",
    "- `vrs240_w18.pro` : Diamond TSV output of RefSeq vs. wolf18\n",
    "\n",
    "### GITB Sequences\n",
    "\n",
    "- `gitb.fa` : High-Confidence, Full-Length RdRp+ RefSeq Sequences\n",
    "\n",
    "\n",
    "- `gitb.rdrp.fa` : RdRp RefSeq sequences\n",
    "```\n",
    "# Header fields:\n",
    "# RefSeq_accession rdrp_start rdrp_end rdrp_len\n",
    "```\n",
    "- `gitb.pro` : Diamond output high-confidence subset of `vrs240_w18.pro` \n",
    "\n",
    "### Cross Validation Sets\n",
    "\n",
    "- `cvi/`\n",
    "\n",
    "- `gitb.all.fa` : same as ../gitb.fa\n",
    "- `gitb.45.cvi`: Cross-validation set at 45% identity (40-50% sequence matchs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## WOLF18 RdRp Header Parsing\n",
    "\n",
    "Sequence data: `ftp://ftp.ncbi.nlm.nih.gov/pub/wolf/_suppl/rnavir18/RNAvirome.S2.afa`\n",
    "Original file: `201230_gitb/wolf18/gb_rdrp.afa`\n",
    "Parsed file: `201230_gitb/wolf18.fa`\n",
    "\n",
    "The headers in this data is pre-processe\n",
    "\n",
    "### Level 1 - Supergroup / Branches\n",
    "\n",
    "The RdRp can be broadly classified into 5 branches which will form the lowest level of the hierarchy: `rdrp1`, `rdrp2`, `rdrp3`, `rdrp4`, `rdrp5`.\n",
    "\n",
    ">Branch 1 consists of leviviruses and their eukaryotic relatives, namely, “mitoviruses,” “narnaviruses,” and “ourmiaviruses” (the latter three terms are placed in quotation marks as our analysis contradicts the current ICTV framework, which classifies mitoviruses and narnaviruses as members of one family, Narnaviridae, and ourmiaviruses as members of a free-floating genus, Ourmiavirus).\n",
    "\n",
    "> Branch 2 (“picornavirus supergroup”) consists of a large assemblage of +RNA viruses of eukaryotes, in particular, those of the orders Picornavirales and Nidovirales; the families Caliciviridae, Potyviridae, Astroviridae, and Solemoviridae, a lineage of dsRNA viruses that includes partitiviruses and picobirnaviruses; and several other, smaller groups of +RNA and dsRNA viruses.\n",
    "\n",
    "> Branch 3 consists of a distinct subset of +RNA viruses, including the “alphavirus supergroup” along with the “flavivirus supergroup,” nodaviruses, and tombusviruses; the “statovirus,” “wèivirus,” “yànvirus,” and “zhàovirus” groups; and several additional, smaller groups.\n",
    "\n",
    "> Branch 4 consists of dsRNA viruses, including cystoviruses, reoviruses, and totiviruses and several additional families.\n",
    "\n",
    "> Branch 5 consists of −RNA viruses.\n",
    "\n",
    "Boundary defintions of Branches with relation to RdRp are taken from paper\n",
    "\n",
    "Based on: [Supplementary Data 4](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6282212/bin/mbo006184203sd4.txt)\n",
    "Saved as: `rdrp_representative_branches.tree`\n",
    "\n",
    "### Level 2 - Viral Family\n",
    "\n",
    "Based on: `https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6282212/bin/mbo006184203sd1.xls`\n",
    "Saved as: `wolf18/wolf18_vlist.xlsx`\n",
    "\n",
    "Spreadsheet includes the fields\n",
    "\n",
    "- RdRp num ID: Ordinal numbering for RdRp\n",
    "- RdRp GenBank Acc: Protein accession ID\n",
    "- NCBI Tax ID: taxid from NCBI\n",
    "- virus name: virus name\n",
    "- taxonomy: taxonomic tree\n",
    "\n",
    "Taxonomy field was parsed to retrieve \"*dae* suffix for \"Family\", relatively appropriate family-name or \"unclassified\" when unavailable. Monkey work.\n",
    "\n",
    "### Level 3 - Sequence/Species\n",
    "\n",
    "Based on: `wolf18_vlist.xlsx`\n",
    "\n",
    "- Virus name field (most shallow taxonomic classification) taken for each record.\n",
    "- GenBank accession taken from each record.\n",
    "\n",
    "### Example RdRp\n",
    "\n",
    "```\n",
    "rdrp5.Hantaviridae.Bowe_virus:AGW23849.1\n",
    "rdrp5.Bunyaviridae.Azagny_virus:AEA42011.1\n",
    "...\n",
    "rdrp2.Coronaviridae.Night_heron_coronavirus_HKU19:YP_005352862.1\n",
    "rdrp2.Coronaviridae.Munia_coronavirus_HKU13_3514:YP_002308505.1\n",
    "rdrp2.Coronaviridae.Wigeon_coronavirus_HKU20:YP_005352870.1\n",
    "rdrp2.Coronaviridae.Feline_infectious_peritonitis_virus:AGZ84535.1\n",
    "rdrp2.Coronaviridae.Lucheng_Rn_rat_coronavirus:YP_009336483.1\n",
    "rdrp2.Coronaviridae.Hipposideros_bat_coronavirus_HKU10:AFU92121.1\n",
    "rdrp2.Coronaviridae.BtMs_AlphaCoV_GS2013:AIA62270.1\n",
    "rdrp2.Coronaviridae.Chaerephon_bat_coronavirus_Kenya_KY41_2006:ADX59465.1\n",
    "rdrp2.Coronaviridae.Porcine_epidemic_diarrhea_virus:AID56804.1\n",
    "rdrp2.Coronaviridae.Bat_coronavirus_CDPHE15_USA_2006:YP_008439200.1\n",
    "rdrp2.Coronaviridae.Anlong_Ms_bat_coronavirus:AID16674.1\n",
    "rdrp2.Coronaviridae.Scotophilus_bat_coronavirus_512:YP_001351683.1\n",
    "rdrp2.Coronaviridae.BtNv_AlphaCoV_SC2013:YP_009201729.1\n",
    "rdrp2.Coronaviridae.Bat_coronavirus_1B:ACA52156.1\n",
    "rdrp2.Coronaviridae.NL63_related_bat_coronavirus:YP_009328933.1\n",
    "rdrp2.Coronaviridae.NL63_related_bat_coronavirus:APD51489.1\n",
    "rdrp2.Coronaviridae.229E_related_bat_coronavirus:ALK43115.1\n",
    "rdrp2.Coronaviridae.Rhinolophus_bat_coronavirus_HKU2:ABQ57223.1\n",
    "rdrp2.Coronaviridae.Wencheng_Sm_shrew_coronavirus:AID16677.1\n",
    "rdrp2.Coronaviridae.Human_coronavirus_HKU1:ABD75543.1\n",
    "rdrp2.Coronaviridae.Betacoronavirus_Erinaceus_VMC_DEU_2012:YP_008719930.1\n",
    "rdrp2.Coronaviridae.Pipistrellus_bat_coronavirus_HKU5:YP_001039961.1\n",
    "rdrp2.Coronaviridae.Rousettus_bat_coronavirus:AOG30811.1\n",
    "rdrp2.Coronaviridae.Rousettus_bat_coronavirus:YP_009273004.1\n",
    "rdrp2.Coronaviridae.Bat_CoV_279_2005:P0C6V9.1\n",
    "rdrp2.Coronaviridae.Bat_Hp_betacoronavirus_Zhejiang2013:YP_009072438.1\n",
    "rdrp2.Coronaviridae.Bottlenose_dolphin_coronavirus_HKU22:AHB63494.1\n",
    "rdrp2.Coronaviridae.Duck_coronavirus:AKF17722.1\n",
    "rdrp2.Coronaviridae.Avian_infectious_bronchitis_virus_partridge_GD_S14_2003:AAT70770.1\n",
    "rdrp2.Coronaviridae.Infectious_bronchitis_virus:ADA83556.1\n",
    "...\n",
    "rdrp1.unclassified.Wenzhou_levi_like_virus_3:APG77299.1\n",
    "rdrp1.Leviviridae.Pseudomonas_phage_PP7:NP_042307.1\n",
    "\n",
    "```\n",
    "\n",
    "saved as: `wolf18/gb_assign_group_v2.txt`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parsing Code\n",
    "\n",
    "# Rename the header to remove virus name\n",
    "# remove gaps from sequence (unaligned)\n",
    "sed 's/|.*//g' gb_rdrp.afa \\\n",
    "  | sed 's/-//g' - \\\n",
    "  > rdrp_1.tmp\n",
    "\n",
    "# One sequence \"Pseudomonas_phage_phiYY\" has no accession\n",
    "# YP_009618381.1\n",
    "sed -i 's/^>$/>YP_009618381.1/g' rdrp_1.tmp\n",
    "\n",
    "# Iterate through each line / fasta name.\n",
    "# to swap out headers\n",
    "while read -r line; do\n",
    "  # Find headers\n",
    "  if [[ \"$line\" = \">\"* ]]; then\n",
    "    acc=$(echo $line | sed 's/\\..*//g' - | sed 's/>//g' -)\n",
    "    newheader=$(grep \"$acc\" gb_assign_group_v2.txt)\n",
    "    \n",
    "    if [[ \"$newheader\" = \"\" ]]; then\n",
    "      echo \">NA_$acc\" >> rdrp_2.tmp\n",
    "    else\n",
    "      echo $newheader >> rdrp_2.tmp\n",
    "    fi\n",
    "    \n",
    "  else\n",
    "    # print aa sequence\n",
    "    echo $line >> rdrp_2.tmp \n",
    "  fi\n",
    "done < rdrp_1.tmp\n",
    "\n",
    "\n",
    "# Manually remove first ten sequences (Group II introns outgroup)\n",
    "tail -n +21 rdrp_2.tmp > rdrp_3.tmp\n",
    "\n",
    "mv rdrp_3.tmp ../wolf18.fa\n",
    "\n",
    "#rm *.tmp"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Viral RefSeq 240\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# RefSeq (rs) folder\n",
    "mkdir -p rs; cd rs\n",
    "\n",
    "wget https://ftp.ncbi.nlm.nih.gov/refseq/release/viral/viral.1.protein.faa.gz\n",
    "wget https://ftp.ncbi.nlm.nih.gov/refseq/release/viral/viral.2.protein.faa.gz\n",
    "gzip -d *.gz\n",
    "\n",
    "cat viral.1.protein.faa viral.2.protein.faa > vrs240.fa\n",
    "rm viral*faa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Run Diamond with ULTRA-SENSITIVE mode\n",
    "QUERY='vrs240.fa'\n",
    "RDRP='wolf18'\n",
    "OUTNAME='vrs240_w18'\n",
    "\n",
    "diamond makedb --in $RDRP.fa -d $RDRP\n",
    "\n",
    "# Diamond blastp alignment\n",
    "time cat $QUERY |\\\n",
    "diamond blastp \\\n",
    "  -d \"$RDRP\".dmnd \\\n",
    "  --unal 0 \\\n",
    "  --masking 0 \\\n",
    "  --ultra-sensitive \\\n",
    "  -k 1 \\\n",
    "  -p 16 \\\n",
    "  -b 1 \\\n",
    "  -f 6 qseqid qstart qend qlen qstrand \\\n",
    "       sseqid sstart send slen \\\n",
    "       pident evalue cigar \\\n",
    "       qseq \\\n",
    "  > \"$OUTNAME\".pro\n",
    "\n",
    "\n",
    "# Fasta of local diamond matches\n",
    "cut -f1,2,3,4,6,7,8,9,10,13 $OUTNAME.pro \\\n",
    "  | sed 's/^/>/g' \\\n",
    "  | sed 's/\\(\\t\\)\\([A-Z]*$\\)/\\n\\2/g' - \\\n",
    "  > $OUTNAME.fa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract High confidence set of matches\n",
    "GITB='gitb'\n",
    "\n",
    "# Extract only matches that are:\n",
    "# >88% identity to rdrp0\n",
    "# >98% coverage of rdrp0\n",
    "\n",
    "cut -f1,2,3,4,6,7,8,9,10,13 $OUTNAME.pro \\\n",
    "  > pro.tmp\n",
    "\n",
    "rm $GITB.pro\n",
    "\n",
    "while read -r line; do\n",
    "  \n",
    "  pctid=$(echo $line | cut -f 9 -d' ' - )  \n",
    "  \n",
    "  subj_start=$(echo $line | cut -f 6 -d' ' - )\n",
    "  subj_end=$(echo $line | cut -f 7 -d' ' - )\n",
    "  subj_len=$(echo $line | cut -f 8 -d' ' - )\n",
    "  \n",
    "  subj_cov=$( echo \"$subj_end - $subj_start + 1\" | bc )\n",
    "  subj_covpct=$( echo \"scale=2;$subj_cov / $subj_len\" | bc )\n",
    "  \n",
    "  if [[ $pctid > 0.88 ]] && [[ $subj_covpct > 0.98 ]]; then\n",
    "    echo $line >> $GITB.pro\n",
    "  fi\n",
    "  \n",
    "done < pro.tmp\n",
    "rm pro.tmp\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parse GITB set into fasta file\n",
    "# Fields:\n",
    "# Query_accession rdrp_start rdrp_end rdrp_len\n",
    "cut -f1,2,3,10 -d' ' $GITB.pro \\\n",
    "  | sed 's/^/>/g' \\\n",
    "  | sed 's/\\( \\)\\([A-Z]*$\\)/\\n\\2/g' - \\\n",
    "  > $GITB.rdrp.fa\n",
    "\n",
    "# rdrp sequences in GITB as bed3\n",
    "cut -f1,2,3 -d' ' $GITB.pro \\\n",
    "  | sed 's/ /\\t/g' - \\\n",
    "  > $GITB.rdrp.bed\n",
    "\n",
    "# From GITB, retrieve full-length records from RefSeq\n",
    "# (complete ORFs)\n",
    "cut -f1 $GITB.rdrp.bed | sed 's/>//g' - > $GITB.hits\n",
    "seqkit grep -r -f $GITB.hits vrs240.fa > $GITB.fa\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Cross-Validation Set of GITB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create Cross-Validation Set\n",
    "mkdir cvi; cd cvi\n",
    "GITB=\"gitb\"\n",
    "\n",
    "# High-Confidence RDRP only\n",
    "ln -s ../gitb.fa ./$GITB.all.fa\n",
    "\n",
    "# USEARCH CVI\n",
    "usearch -threads 30 -calc_distmx $GITB.all.fa -maxdist 1 -termdist 1 -tabbedout gitb.distmat\n",
    "\n",
    "# 90% CVI \n",
    "usearch -distmx_split_identity  gitb.distmat -mindist 0.05 -maxdist 0.15 \\\n",
    "  -tabbedout $GITB.90.cvi\n",
    "\n",
    "# 75% CVI \n",
    "usearch -distmx_split_identity  gitb.distmat -mindist 0.2 -maxdist 0.3 \\\n",
    "  -tabbedout $GITB.75.cvi\n",
    "\n",
    "# 65% CVI \n",
    "usearch -distmx_split_identity  gitb.distmat -mindist 0.3 -maxdist 0.4 \\\n",
    "  -tabbedout $GITB.65.cvi\n",
    "\n",
    "# 55% CVI \n",
    "usearch -distmx_split_identity  gitb.distmat -mindist 0.4 -maxdist 0.5 \\\n",
    "  -tabbedout $GITB.55.cvi\n",
    "\n",
    "# 45% CVI \n",
    "usearch -distmx_split_identity  gitb.distmat -mindist 0.5 -maxdist 0.6 \\\n",
    "  -tabbedout $GITB.45.cvi\n",
    "\n",
    "# 25% CVI \n",
    "usearch -distmx_split_identity gitb.distmat -mindist 0.7 -maxdist 0.8 \\\n",
    "  -tabbedout $GITB.25.cvi\n",
    "  \n",
    "#1 Subset name.\n",
    "#2 Label1\n",
    "#3 Label2\n",
    "#4 Dist"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
