{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# rdrp Benchmark / GITB\n",
    "```\n",
    "Lead     : ababaian /rce\n",
    "Issue    : \n",
    "start    : 2020 12 30\n",
    "complete : 2021 01 01\n",
    "files    : ~/serratus/notebook/20123_gitb/\n",
    "s3 files : s3://serratus-public/notebook/201230_gitb/\n",
    "```\n",
    "\n",
    "## Introduction\n",
    "\n",
    "> `rce:`\n",
    "> For validation of RdRp identification and trimming, we should do hold-out cross-validation using high-quality sequences, i.e. RefSeqs. To assess how accuracy varies with identity, we will use cross-validation by identity (CVI, https://www.drive5.com/usearch/manual/cvi.html) which constructs test/training sets where the maximum identity (MaxID) of the top test-training hit is fixed to a known value or range of values. Here, I suggest five test/training pairs with MaxID = 90, 75, 50, 30, 15% to get a representative range.\n",
    ">\n",
    "> The gold standard reference for the gene identification and trimming benchmark (GITB-gold) will be full-length RdRp-containing ORFs from RefSeqs plus coordinates of the trimmed RdRps. These coordinates will be obtained by aligning the ORFs to Wolf18. These alignments should be good (diamond may be good enough) because the top-hit identities should be at least 90%.\n",
    ">\n",
    "> The decoy reference (GITB-decoy) will be all RefSeq ORFs which do NOT contain RdRp. Ideally, these will all be discarded by the protocol. In practice, there will be FPs due to local homology between RdRp and other genes; this surely arises often in viruses due to recombination events. Using the decoy reference will enable us to design better filters and measure the FP rate (ROC curve) of the protocol.\n",
    "\n",
    "Most of the code for this was developed already in `201210_RdRp_panproteome_v1.ipyn` notebook. This is a self-contained migration of that code / clean-up."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wed Dec 30 15:25:39 PST 2020\n"
     ]
    }
   ],
   "source": [
    "# date and version\n",
    "date"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Upload gitb folder (from EC2)\n",
    "aws s3 sync ./ s3://serratus-public/notebook/201230_gitb/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## EC2 Benchmark environment\n",
    "\n",
    "- OS: `Amazon Linux 2 AMI (HVM) x86`\n",
    "- ami: `ami-0be2609ba883822ec`\n",
    "- instance: `c5.xlarge`\n",
    "- description: `\"c5.xlarge (- ECUs, 4 vCPUs, 3.4 GHz, -, 8 GiB memory, EBS only)\"`\n",
    "- storage: `12 GiB SSD (gp2)`\n",
    "- encryption: `false`\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# From base amazon linux 2\n",
    "sudo yum install -y docker\n",
    "sudo yum install -y git\n",
    "sudo service docker start\n",
    "\n",
    "\n",
    "# Launch \"Serratus-Production\" container environment\n",
    "git clone -b diamond-dev https://github.com/ababaian/serratus.git\n",
    "cd serratus/containers\n",
    "\n",
    "# Build `serratus-align` container\n",
    "# NOTE: diamond v0.9.35.136 Installed\n",
    "./build_containers.sh\n",
    "sudo docker run --rm --entrypoint /bin/bash -it serratus-align:latest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# From `serratus-align` container\n",
    "\n",
    "# Install diamond2 \n",
    "# Libraries for building diamond2\n",
    "yum -y install git gcc gcc-c++ glibc-devel \\\n",
    "  cmake patch automake zlib-devel make\n",
    "\n",
    "# grab latest with fix from Benjamin\n",
    "git clone https://github.com/bbuchfink/diamond.git\n",
    "cd diamond\n",
    "\n",
    "mkdir bin; cd bin\n",
    "cmake ..\n",
    "make -j4\n",
    "cp ./diamond /usr/bin/diamond2\n",
    "#sudo make install"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# From `serratus-align` container\n",
    "mkdir /home/serratus/test; cd /home/serratus/test\n",
    "\n",
    "# Login to your AWS account\n",
    "# aws configure\n",
    "# region: us-east-1\n",
    "\n",
    "# Download rdrp test-database and test-data\n",
    "curl https://serratus-public.s3.amazonaws.com/notebook/201226_rdrp0/rev1/rdrp0_r1.fa > rdrp0.fa\n",
    "curl https://serratus-public.s3.amazonaws.com/test-data/fq-blocks/ERR2756788.fq.00 > ERR2756788.fq.00\n",
    "curl https://serratus-public.s3.amazonaws.com/test-data/fq-blocks/ERR2756788.fq.01 > ERR2756788.fq.01\n",
    "curl https://serratus-public.s3.amazonaws.com/test-data/fq-blocks/ERR2756788.fq.02 > ERR2756788.fq.02\n",
    "curl https://serratus-public.s3.amazonaws.com/test-data/fq-blocks/ERR2756788.fq.03 > ERR2756788.fq.03\n",
    "\n",
    "# diamond1 database\n",
    "diamond makedb  --in rdrp0.fa -d rdrp0.v1\n",
    "diamond2 makedb --in rdrp0.fa -d rdrp0.v2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# NOTE: diamond v0.9.35.136\n",
    "# This amounts to ~ $0.02 / library at production scale\n",
    "run_diamond () {\n",
    "    QUERY=\"$1\"\n",
    "    RDRP=\"$2\"\n",
    "    OUTNAME=\"$3\"\n",
    "    \n",
    "    # Diamond blastp \"SERRATUS PRODUCTION CODE\"\n",
    "    time cat $QUERY |\\\n",
    "    diamond blastx \\\n",
    "        -d \"$RDRP\".dmnd \\\n",
    "        --unal 0 \\\n",
    "        -k 1 \\\n",
    "        -p 1 \\\n",
    "        -b 0.4 \\\n",
    "        -f 6 qseqid sseqid qstart qend qlen sstart send slen pident evalue btop cigar qstrand full_qseq sseq \\\n",
    "        > \"$OUTNAME\".pro\n",
    "\n",
    "}\n",
    "\n",
    "run_set () {\n",
    "    run_diamond ERR2756788.fq.00 rdrp0.v1 b0.v1 & \n",
    "    run_diamond ERR2756788.fq.01 rdrp0.v1 b1.v1 &\n",
    "    run_diamond ERR2756788.fq.02 rdrp0.v1 b2.v1 &\n",
    "    run_diamond ERR2756788.fq.03 rdrp0.v1 b3.v1 &\n",
    "    \n",
    "    echo DONE\n",
    "}\n",
    "\n",
    "run_set\n",
    "\n",
    "# real    3m15.520s\n",
    "# user    3m10.552s\n",
    "# sys     0m4.847s\n",
    "\n",
    "# real    3m15.818s\n",
    "# user    6m21.440s\n",
    "# sys     0m9.486s\n",
    "\n",
    "# real    3m15.819s\n",
    "# user    6m21.440s\n",
    "# sys     0m9.486s\n",
    "\n",
    "# real    3m16.375s\n",
    "# user    3m10.927s\n",
    "# sys     0m4.941s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Diamond2 with Buchfunk optimizations\n",
    "\n",
    "run_diamond2 () {\n",
    "    QUERY=\"$1\"\n",
    "    RDRP=\"$2\"\n",
    "    OUTNAME=\"$3\"\n",
    "    \n",
    "    # Diamond blastp alignment \"BUCHFINK OPTIMIZED CODE\"\n",
    "    time cat $QUERY |\\\n",
    "    diamond2 blastx \\\n",
    "        -d \"$RDRP\".dmnd \\\n",
    "        -c1 -p1 -k1 \\\n",
    "        --more-sensitive \\\n",
    "        --masking 0 \\\n",
    "        --target-indexed \\\n",
    "        > \"$OUTNAME\".pro\n",
    "#         --gapped-filter-evalue 0 \\          \n",
    "}\n",
    "\n",
    "run_diamond2 ERR2756788.fq.00 rdrp0.v2 b0.v2\n",
    "# real    3m54.125s\n",
    "# user    3m45.035s\n",
    "# sys     0m7.377s\n",
    "\n",
    "run_set2 () {\n",
    "    run_diamond2 ERR2756788.fq.00 rdrp0.v2 b0.v2 & \n",
    "    run_diamond2 ERR2756788.fq.01 rdrp0.v2 b1.v2 &\n",
    "    run_diamond2 ERR2756788.fq.02 rdrp0.v2 b2.v2 &\n",
    "    run_diamond2 ERR2756788.fq.03 rdrp0.v2 b3.v2 &\n",
    "    \n",
    "    echo DONE\n",
    "}\n",
    "\n",
    "run_set2\n",
    "\n",
    "# [2]   Exit 137                run_diamond2 ERR2756788.fq.01 rdrp0.v2 b1.v2\n",
    "# real    0m43.325s\n",
    "# user    0m29.316s\n",
    "# sys     0m2.881s\n",
    "\n",
    "# real    5m58.967s\n",
    "# user    5m43.560s\n",
    "# sys     0m4.118s\n",
    "\n",
    "# real    6m13.805s\n",
    "# user    5m53.061s\n",
    "# sys     0m9.679s\n",
    "\n",
    "# real    6m21.922s\n",
    "# user    6m11.612s\n",
    "# sys     0m9.805s\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Diamond2 with Buchfunk optimizations and -fast\n",
    "\n",
    "run_diamond2 () {\n",
    "    QUERY=\"$1\"\n",
    "    RDRP=\"$2\"\n",
    "    OUTNAME=\"$3\"\n",
    "    \n",
    "    # Diamond blastp alignment \"BUCHFINK OPTIMIZED CODE\"\n",
    "    time cat $QUERY |\\\n",
    "    diamond2 blastx \\\n",
    "        -d \"$RDRP\".dmnd \\\n",
    "        -c1 -p1 -k1 \\\n",
    "        --masking 0 \\\n",
    "        --target-indexed \\\n",
    "        > \"$OUTNAME\".pro\n",
    "#         --gapped-filter-evalue 0 \\          \n",
    "}\n",
    "\n",
    "run_set2 () {\n",
    "    run_diamond2 ERR2756788.fq.00 rdrp0.v2 b0.v3 & \n",
    "    run_diamond2 ERR2756788.fq.01 rdrp0.v2 b1.v3 &\n",
    "    run_diamond2 ERR2756788.fq.02 rdrp0.v2 b2.v3 &\n",
    "    run_diamond2 ERR2756788.fq.03 rdrp0.v2 b3.v3 &\n",
    "    \n",
    "    echo DONE\n",
    "}\n",
    "\n",
    "run_set2\n",
    "\n",
    "# CRASH\n",
    "#real    0m26.172s\n",
    "#user    0m22.607s\n",
    "#sys     0m2.757s\n",
    "\n",
    "#real    0m46.678s\n",
    "#user    0m41.974s\n",
    "#sys     0m3.417s\n",
    "\n",
    "#real    0m53.690s\n",
    "#user    0m49.022s\n",
    "#sys     0m3.666s\n",
    "\n",
    "#real    0m53.701s\n",
    "#user    0m49.502s\n",
    "#sys     0m3.401s\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Key Files\n",
    "\n",
    "#### wolf: trimmed set of RdRp true positives\n",
    "\n",
    "- `wolf18.fa` : GenBank rdrp records from wolf18\n",
    "- `wolf20.fa` : Yangshen assembled records from wolf20. Exclusive to wolf18.\n",
    "\n",
    "```\n",
    "# Header fields:\n",
    "# rdrp_branch.viral_family.viral_name:accession\n",
    "```\n",
    "\n",
    "#### RefSeq Curated Viral Sequences\n",
    "\n",
    "- `vrs240.fa` : Viral RefSeq 240 Sequences (aa)\n",
    "- `vrs240_w18.fa` : ALL RefSeq Sequences with a wolf18 match (includes low confidence)\n",
    "- `vrs240_w18.pro` : Diamond TSV output of RefSeq vs. wolf18\n",
    "\n",
    "#### `./gitb` Sequences\n",
    "\n",
    "- `gitb.fa` : High-Confidence & Full-Length RdRp+ RefSeq sequences\n",
    "- `gitb.rdrp.fa` : RdRp RefSeq sequences\n",
    "```\n",
    "# Header fields:\n",
    "# RefSeq_accession rdrp_start rdrp_end rdrp_len\n",
    "```\n",
    "- `gitb.pro` : Diamond output high-confidence subset of `vrs240_w18.pro` \n",
    "\n",
    "#### `decoy/` Sequences\n",
    "\n",
    "- `gitb-decoy.fa` : Curated decoy proteins from human, yeast, fly, ecoli and DNA-virus proteomes\n",
    "- `decoy_w18.pro` : False-positive matches of gitb-decoy.fa to wolf18.fa\n",
    "- `proteome/` : individual proteomes\n",
    "\n",
    "#### `cvi/` Cross Validation Sets (usearch)\n",
    "\n",
    "- `cvi/`\n",
    "- `gitb.all.fa` : same as ../gitb.fa\n",
    "- `gitb.45.cvi`: Cross-validation set at 45% identity (40-50% sequence matchs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## WOLF18 RdRp Header Parsing\n",
    "\n",
    "Sequence data: `ftp://ftp.ncbi.nlm.nih.gov/pub/wolf/_suppl/rnavir18/RNAvirome.S2.afa`\n",
    "Original file: `201230_gitb/wolf18/gb_rdrp.afa`\n",
    "Parsed file: `201230_gitb/wolf18.fa`\n",
    "\n",
    "The headers in this data is pre-processe\n",
    "\n",
    "### Level 1 - Supergroup / Branches\n",
    "\n",
    "The RdRp can be broadly classified into 5 branches which will form the lowest level of the hierarchy: `rdrp1`, `rdrp2`, `rdrp3`, `rdrp4`, `rdrp5`.\n",
    "\n",
    ">Branch 1 consists of leviviruses and their eukaryotic relatives, namely, “mitoviruses,” “narnaviruses,” and “ourmiaviruses” (the latter three terms are placed in quotation marks as our analysis contradicts the current ICTV framework, which classifies mitoviruses and narnaviruses as members of one family, Narnaviridae, and ourmiaviruses as members of a free-floating genus, Ourmiavirus).\n",
    "\n",
    "> Branch 2 (“picornavirus supergroup”) consists of a large assemblage of +RNA viruses of eukaryotes, in particular, those of the orders Picornavirales and Nidovirales; the families Caliciviridae, Potyviridae, Astroviridae, and Solemoviridae, a lineage of dsRNA viruses that includes partitiviruses and picobirnaviruses; and several other, smaller groups of +RNA and dsRNA viruses.\n",
    "\n",
    "> Branch 3 consists of a distinct subset of +RNA viruses, including the “alphavirus supergroup” along with the “flavivirus supergroup,” nodaviruses, and tombusviruses; the “statovirus,” “wèivirus,” “yànvirus,” and “zhàovirus” groups; and several additional, smaller groups.\n",
    "\n",
    "> Branch 4 consists of dsRNA viruses, including cystoviruses, reoviruses, and totiviruses and several additional families.\n",
    "\n",
    "> Branch 5 consists of −RNA viruses.\n",
    "\n",
    "Boundary defintions of Branches with relation to RdRp are taken from paper\n",
    "\n",
    "Based on: [Supplementary Data 4](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6282212/bin/mbo006184203sd4.txt)\n",
    "Saved as: `rdrp_representative_branches.tree`\n",
    "\n",
    "### Level 2 - Viral Family\n",
    "\n",
    "Based on: `https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6282212/bin/mbo006184203sd1.xls`\n",
    "Saved as: `wolf18/wolf18_vlist.xlsx`\n",
    "\n",
    "Spreadsheet includes the fields\n",
    "\n",
    "- RdRp num ID: Ordinal numbering for RdRp\n",
    "- RdRp GenBank Acc: Protein accession ID\n",
    "- NCBI Tax ID: taxid from NCBI\n",
    "- virus name: virus name\n",
    "- taxonomy: taxonomic tree\n",
    "\n",
    "Taxonomy field was parsed to retrieve \"*dae* suffix for \"Family\", relatively appropriate family-name or \"unclassified\" when unavailable. Monkey work.\n",
    "\n",
    "### Level 3 - Sequence/Species\n",
    "\n",
    "Based on: `wolf18_vlist.xlsx`\n",
    "\n",
    "- Virus name field (most shallow taxonomic classification) taken for each record.\n",
    "- GenBank accession taken from each record.\n",
    "\n",
    "### Example RdRp\n",
    "\n",
    "```\n",
    "rdrp5.Hantaviridae.Bowe_virus:AGW23849.1\n",
    "rdrp5.Bunyaviridae.Azagny_virus:AEA42011.1\n",
    "...\n",
    "rdrp2.Coronaviridae.Night_heron_coronavirus_HKU19:YP_005352862.1\n",
    "rdrp2.Coronaviridae.Munia_coronavirus_HKU13_3514:YP_002308505.1\n",
    "rdrp2.Coronaviridae.Wigeon_coronavirus_HKU20:YP_005352870.1\n",
    "rdrp2.Coronaviridae.Feline_infectious_peritonitis_virus:AGZ84535.1\n",
    "rdrp2.Coronaviridae.Lucheng_Rn_rat_coronavirus:YP_009336483.1\n",
    "rdrp2.Coronaviridae.Hipposideros_bat_coronavirus_HKU10:AFU92121.1\n",
    "rdrp2.Coronaviridae.BtMs_AlphaCoV_GS2013:AIA62270.1\n",
    "rdrp2.Coronaviridae.Chaerephon_bat_coronavirus_Kenya_KY41_2006:ADX59465.1\n",
    "rdrp2.Coronaviridae.Porcine_epidemic_diarrhea_virus:AID56804.1\n",
    "rdrp2.Coronaviridae.Bat_coronavirus_CDPHE15_USA_2006:YP_008439200.1\n",
    "rdrp2.Coronaviridae.Anlong_Ms_bat_coronavirus:AID16674.1\n",
    "rdrp2.Coronaviridae.Scotophilus_bat_coronavirus_512:YP_001351683.1\n",
    "rdrp2.Coronaviridae.BtNv_AlphaCoV_SC2013:YP_009201729.1\n",
    "rdrp2.Coronaviridae.Bat_coronavirus_1B:ACA52156.1\n",
    "rdrp2.Coronaviridae.NL63_related_bat_coronavirus:YP_009328933.1\n",
    "rdrp2.Coronaviridae.NL63_related_bat_coronavirus:APD51489.1\n",
    "rdrp2.Coronaviridae.229E_related_bat_coronavirus:ALK43115.1\n",
    "rdrp2.Coronaviridae.Rhinolophus_bat_coronavirus_HKU2:ABQ57223.1\n",
    "rdrp2.Coronaviridae.Wencheng_Sm_shrew_coronavirus:AID16677.1\n",
    "rdrp2.Coronaviridae.Human_coronavirus_HKU1:ABD75543.1\n",
    "rdrp2.Coronaviridae.Betacoronavirus_Erinaceus_VMC_DEU_2012:YP_008719930.1\n",
    "rdrp2.Coronaviridae.Pipistrellus_bat_coronavirus_HKU5:YP_001039961.1\n",
    "rdrp2.Coronaviridae.Rousettus_bat_coronavirus:AOG30811.1\n",
    "rdrp2.Coronaviridae.Rousettus_bat_coronavirus:YP_009273004.1\n",
    "rdrp2.Coronaviridae.Bat_CoV_279_2005:P0C6V9.1\n",
    "rdrp2.Coronaviridae.Bat_Hp_betacoronavirus_Zhejiang2013:YP_009072438.1\n",
    "rdrp2.Coronaviridae.Bottlenose_dolphin_coronavirus_HKU22:AHB63494.1\n",
    "rdrp2.Coronaviridae.Duck_coronavirus:AKF17722.1\n",
    "rdrp2.Coronaviridae.Avian_infectious_bronchitis_virus_partridge_GD_S14_2003:AAT70770.1\n",
    "rdrp2.Coronaviridae.Infectious_bronchitis_virus:ADA83556.1\n",
    "...\n",
    "rdrp1.unclassified.Wenzhou_levi_like_virus_3:APG77299.1\n",
    "rdrp1.Leviviridae.Pseudomonas_phage_PP7:NP_042307.1\n",
    "\n",
    "```\n",
    "\n",
    "saved as: `wolf18/gb_assign_group_v2.txt`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parsing Code\n",
    "\n",
    "# Rename the header to remove virus name\n",
    "# remove gaps from sequence (unaligned)\n",
    "sed 's/|.*//g' gb_rdrp.afa \\\n",
    "  | sed 's/-//g' - \\\n",
    "  > rdrp_1.tmp\n",
    "\n",
    "# One sequence \"Pseudomonas_phage_phiYY\" has no accession\n",
    "# YP_009618381.1\n",
    "sed -i 's/^>$/>YP_009618381.1/g' rdrp_1.tmp\n",
    "\n",
    "# Iterate through each line / fasta name.\n",
    "# to swap out headers\n",
    "while read -r line; do\n",
    "  # Find headers\n",
    "  if [[ \"$line\" = \">\"* ]]; then\n",
    "    acc=$(echo $line | sed 's/\\..*//g' - | sed 's/>//g' -)\n",
    "    newheader=$(grep \"$acc\" gb_assign_group_v2.txt)\n",
    "    \n",
    "    if [[ \"$newheader\" = \"\" ]]; then\n",
    "      echo \">NA_$acc\" >> rdrp_2.tmp\n",
    "    else\n",
    "      echo $newheader >> rdrp_2.tmp\n",
    "    fi\n",
    "    \n",
    "  else\n",
    "    # print aa sequence\n",
    "    echo $line >> rdrp_2.tmp \n",
    "  fi\n",
    "done < rdrp_1.tmp\n",
    "\n",
    "\n",
    "# Manually remove first ten sequences (Group II introns outgroup)\n",
    "tail -n +21 rdrp_2.tmp > rdrp_3.tmp\n",
    "\n",
    "mv rdrp_3.tmp ../wolf18.fa\n",
    "\n",
    "#rm *.tmp"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Viral RefSeq 240\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# RefSeq (rs) folder\n",
    "mkdir -p rs; cd rs\n",
    "\n",
    "wget https://ftp.ncbi.nlm.nih.gov/refseq/release/viral/viral.1.protein.faa.gz\n",
    "wget https://ftp.ncbi.nlm.nih.gov/refseq/release/viral/viral.2.protein.faa.gz\n",
    "gzip -d *.gz\n",
    "\n",
    "cat viral.1.protein.faa viral.2.protein.faa > vrs240.fa\n",
    "rm viral*faa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Run Diamond with ULTRA-SENSITIVE mode\n",
    "QUERY='vrs240.fa'\n",
    "RDRP='wolf18'\n",
    "OUTNAME='vrs240_w18'\n",
    "\n",
    "diamond makedb --in $RDRP.fa -d $RDRP\n",
    "\n",
    "# Diamond blastp alignment\n",
    "time cat $QUERY |\\\n",
    "diamond blastp \\\n",
    "  -d \"$RDRP\".dmnd \\\n",
    "  --unal 0 \\\n",
    "  --masking 0 \\\n",
    "  --ultra-sensitive \\\n",
    "  -k 1 \\\n",
    "  -p 16 \\\n",
    "  -b 1 \\\n",
    "  -f 6 qseqid qstart qend qlen qstrand \\\n",
    "       sseqid sstart send slen \\\n",
    "       pident evalue cigar \\\n",
    "       qseq \\\n",
    "  > \"$OUTNAME\".pro\n",
    "\n",
    "\n",
    "# Fasta of local diamond matches\n",
    "cut -f1,2,3,4,6,7,8,9,10,13 $OUTNAME.pro \\\n",
    "  | sed 's/^/>/g' \\\n",
    "  | sed 's/\\(\\t\\)\\([A-Z]*$\\)/\\n\\2/g' - \\\n",
    "  > $OUTNAME.fa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract High confidence set of matches\n",
    "GITB='gitb'\n",
    "\n",
    "# Extract only matches that are:\n",
    "# >88% identity to rdrp0\n",
    "# >98% coverage of rdrp0\n",
    "\n",
    "cut -f1,2,3,4,6,7,8,9,10,13 $OUTNAME.pro \\\n",
    "  > pro.tmp\n",
    "\n",
    "rm $GITB.pro\n",
    "\n",
    "while read -r line; do\n",
    "  \n",
    "  pctid=$(echo $line | cut -f 9 -d' ' - )  \n",
    "  \n",
    "  subj_start=$(echo $line | cut -f 6 -d' ' - )\n",
    "  subj_end=$(echo $line | cut -f 7 -d' ' - )\n",
    "  subj_len=$(echo $line | cut -f 8 -d' ' - )\n",
    "  \n",
    "  subj_cov=$( echo \"$subj_end - $subj_start + 1\" | bc )\n",
    "  subj_covpct=$( echo \"scale=2;$subj_cov / $subj_len\" | bc )\n",
    "  \n",
    "  if [[ $pctid > 0.88 ]] && [[ $subj_covpct > 0.98 ]]; then\n",
    "    echo $line >> $GITB.pro\n",
    "  fi\n",
    "  \n",
    "done < pro.tmp\n",
    "rm pro.tmp\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parse GITB set into fasta file\n",
    "# Fields:\n",
    "# Query_accession rdrp_start rdrp_end rdrp_len\n",
    "cut -f1,2,3,10 -d' ' $GITB.pro \\\n",
    "  | sed 's/^/>/g' \\\n",
    "  | sed 's/\\( \\)\\([A-Z]*$\\)/\\n\\2/g' - \\\n",
    "  > $GITB.rdrp.fa\n",
    "\n",
    "# rdrp sequences in GITB as bed3\n",
    "cut -f1,2,3 -d' ' $GITB.pro \\\n",
    "  | sed 's/ /\\t/g' - \\\n",
    "  > $GITB.rdrp.bed\n",
    "\n",
    "# From GITB, retrieve full-length records from RefSeq\n",
    "# (complete ORFs)\n",
    "cut -f1 $GITB.rdrp.bed | sed 's/>//g' - > $GITB.hits\n",
    "seqkit grep -r -f $GITB.hits vrs240.fa > $GITB.fa\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Cross-Validation Sets of GITB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create Cross-Validation Set\n",
    "mkdir cvi; cd cvi\n",
    "GITB=\"gitb\"\n",
    "\n",
    "# High-Confidence RDRP only\n",
    "ln -s ../gitb.rdrp.fa ./$GITB.rdrp.fa\n",
    "\n",
    "# USEARCH CVI\n",
    "usearch -threads 30 -calc_distmx $GITB.rdrp.fa -maxdist 1 -termdist 1 -tabbedout gitb.distmat\n",
    "\n",
    "# 90% CVI \n",
    "usearch -distmx_split_identity  gitb.distmat -mindist 0.05 -maxdist 0.15 \\\n",
    "  -tabbedout $GITB.90.cvi\n",
    "\n",
    "# 75% CVI \n",
    "usearch -distmx_split_identity  gitb.distmat -mindist 0.2 -maxdist 0.3 \\\n",
    "  -tabbedout $GITB.75.cvi\n",
    "\n",
    "# 65% CVI \n",
    "usearch -distmx_split_identity  gitb.distmat -mindist 0.3 -maxdist 0.4 \\\n",
    "  -tabbedout $GITB.65.cvi\n",
    "\n",
    "# 55% CVI \n",
    "usearch -distmx_split_identity  gitb.distmat -mindist 0.4 -maxdist 0.5 \\\n",
    "  -tabbedout $GITB.55.cvi\n",
    "\n",
    "# 45% CVI \n",
    "usearch -distmx_split_identity  gitb.distmat -mindist 0.5 -maxdist 0.6 \\\n",
    "  -tabbedout $GITB.45.cvi\n",
    "  \n",
    "# 35% CVI \n",
    "usearch -distmx_split_identity  gitb.distmat -mindist 0.6 -maxdist 0.7 \\\n",
    "  -tabbedout $GITB.35.cvi\n",
    "\n",
    "# 25% CVI \n",
    "usearch -distmx_split_identity gitb.distmat -mindist 0.7 -maxdist 0.8 \\\n",
    "  -tabbedout $GITB.25.cvi\n",
    "  \n",
    "#1 Subset name.\n",
    "#2 Label1\n",
    "#3 Label2\n",
    "#4 Dist"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# gitb-decoy set (v0)\n",
    "\n",
    "Annotated set of non RdRp containing proteins as a negative control\n",
    "\n",
    "### Human Proteome\n",
    "File: `homo.fa`\n",
    "UNIPROT: [UP000005640](https://www.uniprot.org/proteomes/UP000005640)\n",
    "Accessed: `201230`\n",
    "Results: `548 843`\n",
    "\n",
    "### Yeast Proteome\n",
    "File: `yeast.fa`\n",
    "UNIPROT: [UP000002311](https://www.uniprot.org/proteomes/UP000002311)\n",
    "Accessed: `201230`\n",
    "Results: `57 973`\n",
    "\n",
    "### Ecoli Proteome\n",
    "File: `ecoli.fa`\n",
    "UNIPROT: [UP000000558](https://www.uniprot.org/proteomes/UP000000558)\n",
    "Accessed: `201230`\n",
    "Results: `33 586`\n",
    "\n",
    "### Monodnaviria Proteome (ssDNA viruses)\n",
    "File: `gb_dsDNA.seq` # Only downloaded accession list\n",
    "GenBank Query: `txid2731342[Organism:exp]`\n",
    "Results: `237452`\n",
    "\n",
    "### Duplodnaviria Proteome (dsDNA viruses)\n",
    "File: `gb_ssDNA.seq`\n",
    "GenBank Query: `txid2731341[Organism:exp]`\n",
    "Results: `1813000`\n",
    "\n",
    "### Varidnaviria Proteome (vDNA viruses)\n",
    "File: `vDNA.fa`\n",
    "GenBank Query: `txid2732004[Organism:exp]`\n",
    "Accessed: `201230`\n",
    "Results: `376693`\n",
    "\n",
    "### Artverviricota  Proteome (Retroviruses)\n",
    "File: `rv.fa`\n",
    "GenBank Query: `txid2732409[Organism:exp] `\n",
    "Accessed: `210105`\n",
    "Results: `531300`\n",
    "\n",
    "### Reverse Transcriptase PFAM \n",
    "File: `rt_pfam.id97.fa` \n",
    "Query: From RCE, 97% clustered\n",
    "Accessed: `210106`\n",
    "Results: `58913`\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create folder for single proteomes\n",
    "mkdir proteome\n",
    "mv rv.fa proteome/\n",
    "mv vdna.fa proteome/\n",
    "mv homo.fa proteome/\n",
    "mv yeast.fa proteome/\n",
    "mv ecoli.fa proteome/\n",
    "\n",
    "### make gitb-decoy.fa\n",
    "\n",
    "# Extract DNA virus accessions from genbank v240\n",
    "cat gb_dsDNA.seq gb_ssDNA.seq gb_vDNA.seq > vdna.acc ; rm *seq\n",
    "seqkit grep -j 30 -f vdna.acc aaViro_gb240.fa > vdna.tmp\n",
    "cat rv.fa vdna.tmp > vdna.fa # RV added later\n",
    "\n",
    "cat vdna.fa homo.fa yeast.fa ecoli.fa rt_pfam.id97.fa \\\n",
    "  | seqkit rmdup -s - \\\n",
    "  | seqkit seq -m 15 \\\n",
    "  > gitb-decoy.fa\n",
    "  \n",
    "# Deplete mis-annotations (see below)\n",
    "echo -e \"CZQ50745.1\\nAFH02745.1\\nAFH02746.1\\nADO67072.1\\npdb|4TN2|A\\nYP_009506261.1\\nADE61677.1\\nYP_009551602.1\\nAXY66749.1\\nAWS06671.1\" > misannotations.tmp\n",
    "\n",
    "seqkit grep -j 30 -v -f misannotations.tmp gitb-decoy.fa > gitb-decoy.fa2\n",
    "rm misannotations.tmp; mv gitb-decoy.fa2 gitb-decoy.fa\n",
    "\n",
    "seqkit faidx -f gitb-decoy.fa\n",
    "md5sum gitb-decoy.fa > gitb-decoy.fa.md5"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Manual clean-up of misannotations\n",
    "\n",
    "- `CZQ50745.1` Pepper leafroll virus is annotated as a ssDNA virus [with an rdrp](https://apsjournals.apsnet.org/doi/pdf/10.1094/PDIS-03-17-0418-RE). Manually remove accession.\n",
    "```\n",
    "CZQ50745.1      655     1085    1085    +       rdrp2.Luteoviridae.Pepper_vein_yellows_virus:AKS03434   2       432     432     96.1    3.9e-301      840.1    431M    VGPQAEVTSLTLQAERWLQRAQSAKIPSSEDRERVINKTVKAYSNVKTFGPTATRGNKLEWRQFLEDFKSAVFSLELDAGIGVPYIAYGRPTHKGWVEDPKLLPVLARLTFNRLQKMLEVESTDMSAEELVQAGLCDPIRTFVKREPHKQSKLDEGRYRLIMSVSLVDQLVARVLFQNQNKREIALWRANPSKPGFGLSTDEQVLEFVQALAAQVEVPPEELITSWKKYLVPTDCSGFDWSVAEWMLHDDMVVRNKLTLDLNPTTEKLRSAWLKCICNSVLCLSDGTLLAQRVPGVQKSGSYNTSSSNSRIRVMAAYHCGADWAMAMGDDALESVNTNLEVYKSLGFKVEVSGQLEFCSHIFRAPDLALPVNERKMLYKLIFGYNPGCGNLEVITNYIAACVSVLNELRHDPDSVALLHTWLVSPVLPQNN\n",
    "```\n",
    "\n",
    "- `AFH02745.1`, `AFH02746.1` and `ADO67072.1`: is annotated as a Reverse Transcriptase from a transposable element in a piccovirus (not picorna like in the header). InterPro predicted an RDRP_1 at coordintes 14-91, this is a mis-annotation, removal justified. also: \n",
    "\n",
    "```\n",
    "AFH02746.1      1       89      95      +       rdrp2.unclassified.Hubei_arthropod_virus_1:YP_009336629 116     204     478     67.4    2.4e-35 122.9 89M      KEQLRKEGIVPETIFVDTLKDERKRSDKIKKQGATRVFCASPVDYTIALRQSMLHFCAAYMKNRHTQMHAVGIDVHAEEWARLWSRLTK\n",
    "```\n",
    "\n",
    "```\n",
    ">AFH02745.1 polyprotein, partial [Bat feces associated picorna-like virus SC2797]\n",
    "QCEEVDHSSLEGAKFIFDEECQVEYIGTVPPNLQPHIPQKTKIEKSLLHTRFGFKSMTEPTILSPQDPRY\n",
    "THELSPLYYGARKHGIKTKDFTSTMVKRAKEALWTRWLSGMEPIVVDPKKLTPEEAVLGFVGNKYYEALS\n",
    "LNTSAGYPWSLGGQTTKDSWIETNWEQRTCVIHPQLLKEIQRKEQLRKEGI\n",
    ">AFH02746.1 polyprotein, partial [Bat feces associated picorna-like virus SC3066]\n",
    "KEQLRKEGIVPETIFVDTLKDERKRSDKIKKQGATRVFCASPVDYTIALRQSMLHFCAAYMKNRHTQMHA\n",
    "VGIDVHAEEWARLWSRLTKHGYKFI\n",
    "\n",
    "```\n",
    "----\n",
    "\n",
    "- In the rv-decoy set (a later addition)\n",
    "- `pdb|4TN2|A`, `YP_009506261.1`, `ADE61677.1`, `YP_009551602.1`, `AXY66749.1`, and `AWS06671.1\n",
    "\n",
    "----\n",
    "- `A0A0B4J238` T cell receptor alpha variable 1-2 Protein. IG-like domain, this is the highest scoring False Positive match at E-value of 4.3e-07\n",
    "\n",
    "```\n",
    "sp|A0A0B4J238|TVA12_HUMAN       9       105     106     +       rdrp5.unclassified.Zirqa_virus:AMT75437 344     450     660     29.9    4.3e-07 44.3  18M2D10M1D45M7D24M       VSMKMGGTTGQNIDQPTEMTATEGAIVQINCTYQTSGFNGLFWYQQHAGEAPTFLSYNVLDGLEEKGRFSSFLSRSKGYSYLLLKELQMKDSASYLC\n",
    "```\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Run Diamond with ULTRA-SENSITIVE mode\n",
    "QUERY='gitb-decoy.fa'\n",
    "RDRP='wolf18'\n",
    "OUTNAME='gitb-decoy_w18'\n",
    "\n",
    "cp ../$RDRP.fa ./\n",
    "diamond makedb --in $RDRP.fa -d $RDRP\n",
    "\n",
    "# Diamond blastp alignment\n",
    "time cat $QUERY |\\\n",
    "diamond blastp \\\n",
    "  -d \"$RDRP\".dmnd \\\n",
    "  --unal 0 \\\n",
    "  --masking 0 \\\n",
    "  --ultra-sensitive \\\n",
    "  -e 0.05 \\\n",
    "  -k 1 \\\n",
    "  -p 16 \\\n",
    "  -b 1 \\\n",
    "  -f 6 qseqid qstart qend qlen qstrand \\\n",
    "       sseqid sstart send slen \\\n",
    "       pident evalue bitscore \\\n",
    "       cigar  qseq \\\n",
    "  > \"$OUTNAME\".pro\n",
    "  \n",
    "# wc -l\n",
    "# 7478 / 349304 false-positive hits in decoy set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aws s3 sync ./ s3://serratus-public/notebook/201230_gitb/decoy/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Biological Cross Validation\n",
    "\n",
    "A fairly simple experiment. Using the `wolf18` rdrp-set as a database, compare the distribution matches froma real set of novel rdrp in `wolf20` vs the `gitb-decoy.fa` and `gitb-yoced.fa` (simulated noise from reverse-decoy) set. In contrast to leave-one-out validation, this probably captures what the SRA will look like for diamond quite well.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mkdir bcv\n",
    "\n",
    "ln -s ../wolf18.fa ./\n",
    "aws s3 cp s3://serratus-public/notebook/201210_rdrp0/yaRdRp_201212.fa ./wolf20.fa\n",
    "ln -s ../decoy/gitb-decoy.fa ./\n",
    "seqkit seq -r gitb-decoy.fa > gitb-yoced.fa\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Run Diamond with ULTRA-SENSITIVE mode\n",
    "\n",
    "run_diamond () {\n",
    "    QUERY=\"$1\"\n",
    "    RDRP=\"$2\"\n",
    "    OUTNAME=\"$3\"\n",
    "\n",
    "    diamond makedb --in $RDRP.fa -d $RDRP\n",
    "\n",
    "    # Diamond blastp alignment\n",
    "    time cat $QUERY |\\\n",
    "    diamond blastp \\\n",
    "      -d \"$RDRP\".dmnd \\\n",
    "      --unal 0 \\\n",
    "      --masking 0 \\\n",
    "      --ultra-sensitive \\\n",
    "      -e 0.05 \\\n",
    "      -k 1 \\\n",
    "      -p 16 \\\n",
    "      -b 1 \\\n",
    "      -f 6 qseqid qstart qend qlen qstrand \\\n",
    "           sseqid sstart send slen \\\n",
    "           pident evalue bitscore \\\n",
    "           cigar  qseq \\\n",
    "      > \"$OUTNAME\".pro\n",
    "}\n",
    "\n",
    "# Diamond Benchmarks\n",
    "run_diamond 'wolf20.fa' 'wolf18' 'w20_w18'\n",
    "run_diamond 'gitb-decoy.fa' 'wolf18' 'decoy_w18'\n",
    "run_diamond 'gitb-yoced.fa' 'wolf18' 'yoced_w18'\n",
    "\n",
    "# upload\n",
    "rm *.fa *.dmnd\n",
    "aws s3 sync ./ s3://serratus-public/notebook/201230_gitb/bcv/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Run Diamond with ULTRA-SENSITIVE - PAM250\n",
    "\n",
    "run_diamond () {\n",
    "    QUERY=\"$1\"\n",
    "    RDRP=\"$2\"\n",
    "    OUTNAME=\"$3\"\n",
    "\n",
    "    diamond makedb --in $RDRP.fa -d $RDRP\n",
    "\n",
    "    # Diamond blastp alignment\n",
    "    time cat $QUERY |\\\n",
    "    diamond blastp \\\n",
    "      -d \"$RDRP\".dmnd \\\n",
    "      --unal 0 \\\n",
    "      --masking 0 \\\n",
    "      --ultra-sensitive \\\n",
    "      --matrix PAM250 \\\n",
    "      -e 0.05 \\\n",
    "      -k 1 \\\n",
    "      -p 16 \\\n",
    "      -b 1 \\\n",
    "      -f 6 qseqid qstart qend qlen qstrand \\\n",
    "           sseqid sstart send slen \\\n",
    "           pident evalue bitscore \\\n",
    "           cigar  qseq \\\n",
    "      > \"$OUTNAME\".pro\n",
    "}\n",
    "\n",
    "# Diamond Benchmarks\n",
    "run_diamond 'wolf20.fa' 'wolf18' 'w20_w18_pam250'\n",
    "run_diamond 'gitb-decoy.fa' 'wolf18' 'decoy_w18_pam250'\n",
    "run_diamond 'gitb-yoced.fa' 'wolf18' 'yoced_w18_pam250'\n",
    "\n",
    "mkdir pam250\n",
    "mv *_pam250.pro pam250/\n",
    "\n",
    "# upload\n",
    "rm *.fa *.dmnd\n",
    "aws s3 sync ./ s3://serratus-public/notebook/201230_gitb/bcv/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Target Optimized Diamond (buchfunk)\n",
    "\n",
    "> `diamond blastx -q query.fna -d rdrp0 -o out -c1 -p1 -k1 --more-sensitive --log --min-orf 1 --masking 0 --gapped-filter-evalue 0 --target-indexed`\n",
    ">\n",
    "> `bb`: The optimization is based on fitting a hash set for the database seeds into the cache, so it may suddenly stop working if the database gets bigger. I would probably in any case recommend using a clustered database (choose id% so you don't lose sensitivity), then do a second round of diamond of the aligned reads against the full database.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Update Diamond"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd gitb/bcv\n",
    "\n",
    "ln -s ../wolf18.fa ./\n",
    "aws s3 cp s3://serratus-public/notebook/201210_rdrp0/yaRdRp_201212.fa ./wolf20.fa\n",
    "ln -s ../decoy/gitb-decoy.fa ./\n",
    "seqkit seq -r gitb-decoy.fa > gitb-yoced.fa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Run Diamond with/without buchfink opt\n",
    "\n",
    "run_diamond () {\n",
    "    QUERY=\"$1\"\n",
    "    RDRP=\"$2\"\n",
    "    OUTNAME=\"$3\"\n",
    "\n",
    "    diamond makedb --in $RDRP.fa -d $RDRP\n",
    "\n",
    "# Diamond blastp alignment\n",
    "    time cat $QUERY |\\\n",
    "    diamond blastp \\\n",
    "        -d \"$RDRP\".dmnd \\\n",
    "        -c1 -p1 -k1 -e 0.05 \\\n",
    "        --more-sensitive \\\n",
    "        --masking 0 \\\n",
    "        --gapped-filter-evalue 0 \\\n",
    "        --target-indexed \\\n",
    "        > \"$OUTNAME\".pro\n",
    "}\n",
    "\n",
    "# Diamond Benchmarks\n",
    "run_diamond 'wolf20.fa' 'wolf18' 'w20_w18_bb'\n",
    "run_diamond 'gitb-decoy.fa' 'wolf18' 'decoy_w18_bb'\n",
    "run_diamond 'gitb-yoced.fa' 'wolf18' 'yoced_w18_bb'\n",
    "\n",
    "mkdir pam250\n",
    "mv *_pam250.pro pam250/\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
